// news-extractor-node/dist/AuthorExtractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthorExtractor = void 0;
class AuthorExtractor {
    constructor() {
        this.authorPatterns = [
            /作者[：:]\s*(.+)/,
            /记者[：:]\s*(.+)/,
            /编辑[：:]\s*(.+)/,
            /撰稿[：:]\s*(.+)/,
            /By\s+(.+)/i,
            /Author[：:]\s*(.+)/i
        ];
        this.authorSelectors = [
            '.author',
            '.article-author',
            '.post-author',
            '[rel="author"]',
            'meta[name="author"]',
            'meta[property="article:author"]'
        ];
    }
    extract(dom, authorXPath) {
        const document = dom.window.document;
        // If custom XPath provided, use it
        if (authorXPath) {
            const result = this.evaluateXPath(document, authorXPath);
            if (result) {
                return result.trim();
            }
        }
        // Try common selectors
        for (const selector of this.authorSelectors) {
            const element = document.querySelector(selector);
            if (element) {
                // Check content attribute (for meta tags)
                const content = element.getAttribute('content');
                if (content) {
                    return this.cleanAuthorName(content);
                }
                // Check text content
                const text = element.textContent;
                if (text) {
                    const author = this.cleanAuthorName(text);
                    if (author && author.length < 50) {
                        return author;
                    }
                }
            }
        }
        // Search in body text
        const bodyText = document.body?.textContent || '';
        return this.extractAuthorFromText(bodyText);
    }
    extractAuthorFromText(text) {
        for (const pattern of this.authorPatterns) {
            const match = text.match(pattern);
            if (match && match[1]) {
                const author = this.cleanAuthorName(match[1]);
                if (author.length < 50) {
                    return author;
                }
            }
        }
        return '';
    }
    cleanAuthorName(name) {
        // Remove common prefixes/suffixes
        name = name.trim();
        name = name.replace(/^(作者|记者|编辑|撰稿|By|Author)[：:\s]+/i, '');
        name = name.replace(/\s*(发布|编辑|整理)$/, '');
        // Remove extra whitespace
        name = name.replace(/\s+/g, ' ');
        // Take first line if multiple lines
        name = name.split('\n')[0];
        return name.trim();
    }
    evaluateXPath(document, xpath) {
        try {
            const result = document.evaluate(xpath, document, null, 2, // STRING_TYPE
            null);
            return result.stringValue;
        }
        catch (e) {
            return '';
        }
    }
}
exports.AuthorExtractor = AuthorExtractor;
