// news-extractor-node/dist/ContentExtractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentExtractor = void 0;
class ContentExtractor {
    constructor() {
        this.punctuation = new Set([
            '！', '，', '。', '？', '、', '；', '：', '\u201c', '\u201d', '\u2018', '\u2019',
            '《', '》', '%', '（', '）', ',', '.', '?', ':', ';', '!', '%', '(', ')'
        ]);
    }
    extract(dom, contentXPath) {
        const document = dom.window.document;
        // If custom XPath provided, use it
        if (contentXPath) {
            const result = this.evaluateXPath(document, contentXPath);
            if (result) {
                return {
                    text: result.textContent || '',
                    html: result.innerHTML || '',
                    images: this.extractImages(result)
                };
            }
        }
        // Otherwise, use text density algorithm
        const body = document.body;
        if (!body) {
            return { text: '', html: '', images: [] };
        }
        const nodeInfoMap = new Map();
        this.calculateTextDensity(body, nodeInfoMap);
        // Find the node with highest text density
        let maxDensity = 0;
        let bestNode = null;
        nodeInfoMap.forEach((info, node) => {
            if (info.density > maxDensity && info.textCount > 100) {
                maxDensity = info.density;
                bestNode = node;
            }
        });
        if (!bestNode) {
            return { text: '', html: '', images: [] };
        }
        const element = bestNode;
        return {
            text: element.textContent || '',
            html: element.innerHTML || '',
            images: this.extractImages(element)
        };
    }
    calculateTextDensity(node, nodeInfoMap) {
        if (nodeInfoMap.has(node)) {
            return nodeInfoMap.get(node);
        }
        let textCount = 0;
        let tagCount = 0;
        // Count direct text nodes
        for (const child of Array.from(node.childNodes)) {
            if (child.nodeType === 3) { // Text node
                const text = child.textContent || '';
                textCount += this.countEffectiveText(text);
            }
            else if (child.nodeType === 1) { // Element node
                tagCount++;
                const childInfo = this.calculateTextDensity(child, nodeInfoMap);
                textCount += childInfo.textCount;
                tagCount += childInfo.tagCount;
            }
        }
        const density = tagCount === 0 ? 0 : textCount / tagCount;
        const info = {
            textCount,
            tagCount,
            density,
            text: node.textContent || '',
            html: node.innerHTML || ''
        };
        nodeInfoMap.set(node, info);
        return info;
    }
    countEffectiveText(text) {
        // Remove whitespace
        const cleaned = text.replace(/\s+/g, '');
        // Count characters, giving more weight to punctuation
        let count = 0;
        for (const char of cleaned) {
            if (this.punctuation.has(char)) {
                count += 5; // Punctuation has higher weight
            }
            else {
                count += 1;
            }
        }
        return count;
    }
    extractImages(element) {
        const images = [];
        const imgElements = element.querySelectorAll('img');
        imgElements.forEach(img => {
            const src = img.getAttribute('src');
            if (src) {
                images.push(src);
            }
        });
        return images;
    }
    evaluateXPath(document, xpath) {
        try {
            const result = document.evaluate(xpath, document, null, 9, // FIRST_ORDERED_NODE_TYPE
            null);
            return result.singleNodeValue;
        }
        catch (e) {
            return null;
        }
    }
}
exports.ContentExtractor = ContentExtractor;
