// news-extractor-node/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewsExtractor = void 0;
exports.extractNews = extractNews;
const jsdom_1 = require("jsdom");
const ContentExtractor_1 = require("./ContentExtractor");
const TitleExtractor_1 = require("./TitleExtractor");
const TimeExtractor_1 = require("./TimeExtractor");
const AuthorExtractor_1 = require("./AuthorExtractor");
class NewsExtractor {
    constructor() {
        this.contentExtractor = new ContentExtractor_1.ContentExtractor();
        this.titleExtractor = new TitleExtractor_1.TitleExtractor();
        this.timeExtractor = new TimeExtractor_1.TimeExtractor();
        this.authorExtractor = new AuthorExtractor_1.AuthorExtractor();
    }
    extract(html, options = {}) {
        // Parse HTML
        const dom = new jsdom_1.JSDOM(html);
        const document = dom.window.document;
        // Remove noise elements
        if (options.noiseSelectors) {
            this.removeNoiseElements(document, options.noiseSelectors);
        }
        // Extract metadata
        const title = this.titleExtractor.extract(dom, options.titleXPath);
        const author = this.authorExtractor.extract(dom, options.authorXPath);
        const publishTime = this.timeExtractor.extract(dom, options.publishTimeXPath);
        // Extract content
        const contentResult = this.contentExtractor.extract(dom, options.contentXPath);
        // Resolve relative URLs for images
        const images = options.url
            ? contentResult.images.map(img => this.resolveUrl(img, options.url))
            : contentResult.images;
        return {
            title,
            author,
            publishTime,
            content: contentResult.text,
            contentHtml: contentResult.html,
            images
        };
    }
    removeNoiseElements(document, selectors) {
        for (const selector of selectors) {
            const elements = document.querySelectorAll(selector);
            elements.forEach(el => el.remove());
        }
    }
    resolveUrl(url, baseUrl) {
        try {
            return new URL(url, baseUrl).href;
        }
        catch (e) {
            return url;
        }
    }
}
exports.NewsExtractor = NewsExtractor;
// Export types
__exportStar(require("./types"), exports);
// Convenience function
function extractNews(html, options = {}) {
    const extractor = new NewsExtractor();
    return extractor.extract(html, options);
}
