// news-extractor-node/dist/TimeExtractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeExtractor = void 0;
class TimeExtractor {
    constructor() {
        this.timePatterns = [
            // ISO 8601 format
            /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/,
            // Common date formats
            /\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}/,
            /\d{4}\/\d{2}\/\d{2}\s+\d{2}:\d{2}:\d{2}/,
            /\d{4}年\d{1,2}月\d{1,2}日\s*\d{1,2}:\d{2}/,
            /\d{4}-\d{2}-\d{2}/,
            /\d{4}\/\d{2}\/\d{2}/,
            /\d{4}年\d{1,2}月\d{1,2}日/
        ];
        this.timeSelectors = [
            'time',
            '[datetime]',
            '.publish-time',
            '.post-time',
            '.article-time',
            'meta[property="article:published_time"]',
            'meta[name="pubdate"]'
        ];
    }
    extract(dom, publishTimeXPath) {
        const document = dom.window.document;
        // If custom XPath provided, use it
        if (publishTimeXPath) {
            const result = this.evaluateXPath(document, publishTimeXPath);
            if (result) {
                return result.trim();
            }
        }
        // Try common selectors
        for (const selector of this.timeSelectors) {
            const element = document.querySelector(selector);
            if (element) {
                // Check datetime attribute
                const datetime = element.getAttribute('datetime');
                if (datetime) {
                    return this.normalizeTime(datetime);
                }
                // Check content attribute (for meta tags)
                const content = element.getAttribute('content');
                if (content) {
                    return this.normalizeTime(content);
                }
                // Check text content
                const text = element.textContent;
                if (text) {
                    const time = this.extractTimeFromText(text);
                    if (time) {
                        return time;
                    }
                }
            }
        }
        // Search in body text
        const bodyText = document.body?.textContent || '';
        return this.extractTimeFromText(bodyText) || '';
    }
    extractTimeFromText(text) {
        for (const pattern of this.timePatterns) {
            const match = text.match(pattern);
            if (match) {
                return this.normalizeTime(match[0]);
            }
        }
        return '';
    }
    normalizeTime(timeStr) {
        // Remove extra whitespace
        timeStr = timeStr.trim();
        // Try to parse and format
        try {
            const date = new Date(timeStr);
            if (!isNaN(date.getTime())) {
                return date.toISOString().split('T')[0] + ' ' +
                    date.toTimeString().split(' ')[0];
            }
        }
        catch (e) {
            // Ignore parsing errors
        }
        return timeStr;
    }
    evaluateXPath(document, xpath) {
        try {
            const result = document.evaluate(xpath, document, null, 2, // STRING_TYPE
            null);
            return result.stringValue;
        }
        catch (e) {
            return '';
        }
    }
}
exports.TimeExtractor = TimeExtractor;
