// @siping/html-to-markdown-node/dist/utils/dom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBlockElement = isBlockElement;
exports.isInlineElement = isInlineElement;
exports.getNodeName = getNodeName;
exports.isElement = isElement;
exports.isTextNode = isTextNode;
exports.getTextContent = getTextContent;
exports.hasChildNodes = hasChildNodes;
exports.getAttribute = getAttribute;
const BLOCK_ELEMENTS = new Set([
    'address', 'article', 'aside', 'blockquote', 'details', 'dialog', 'dd', 'div',
    'dl', 'dt', 'fieldset', 'figcaption', 'figure', 'footer', 'form', 'h1', 'h2',
    'h3', 'h4', 'h5', 'h6', 'header', 'hgroup', 'hr', 'li', 'main', 'nav', 'ol',
    'p', 'pre', 'section', 'table', 'ul', 'tbody', 'thead', 'tfoot', 'tr', 'td', 'th'
]);
const INLINE_ELEMENTS = new Set([
    'a', 'abbr', 'b', 'bdi', 'bdo', 'br', 'cite', 'code', 'data', 'dfn', 'em',
    'i', 'kbd', 'mark', 'q', 'rp', 'rt', 'ruby', 's', 'samp', 'small', 'span',
    'strong', 'sub', 'sup', 'time', 'u', 'var', 'wbr', 'del', 'ins', 'strike'
]);
function isBlockElement(tagName) {
    return BLOCK_ELEMENTS.has(tagName.toLowerCase());
}
function isInlineElement(tagName) {
    return INLINE_ELEMENTS.has(tagName.toLowerCase());
}
function getNodeName(node) {
    if (node.nodeType === 1) {
        return node.tagName.toLowerCase();
    }
    return '';
}
function isElement(node) {
    return node.nodeType === 1;
}
function isTextNode(node) {
    return node.nodeType === 3;
}
function getTextContent(node) {
    return node.textContent || '';
}
function hasChildNodes(node) {
    return node.childNodes && node.childNodes.length > 0;
}
function getAttribute(node, attr) {
    if (isElement(node)) {
        return node.getAttribute(attr);
    }
    return null;
}
