// @siping/html-to-markdown-node/dist/cli.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = main;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const index_1 = require("./index");
function parseArgs(args) {
    const options = {};
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        switch (arg) {
            case '-h':
            case '--help':
                options.help = true;
                break;
            case '-v':
            case '--version':
                options.version = true;
                break;
            case '-i':
            case '--input':
                options.input = args[++i];
                break;
            case '-o':
            case '--output':
                options.output = args[++i];
                break;
            case '-d':
            case '--domain':
                options.domain = args[++i];
                break;
            case '--escape-mode':
                options.escapeMode = args[++i];
                break;
        }
    }
    return options;
}
function showHelp() {
    console.log(`
html-to-markdown-node - Convert HTML to Markdown

Usage:
  html2md [options]
  echo "<html>" | html2md
  html2md -i input.html -o output.md

Options:
  -i, --input <file>        Input HTML file (default: stdin)
  -o, --output <file>       Output Markdown file (default: stdout)
  -d, --domain <url>        Base domain for resolving relative URLs
  --escape-mode <mode>      Escape mode: 'smart' or 'disabled' (default: smart)
  -h, --help                Show this help message
  -v, --version             Show version number

Examples:
  # Convert from stdin to stdout
  echo "<strong>Bold</strong>" | html2md

  # Convert file
  html2md -i input.html -o output.md

  # With domain for URL resolution
  html2md -i input.html -d https://example.com

  # Disable escaping
  html2md -i input.html --escape-mode disabled
`);
}
function showVersion() {
    const packageJson = require('../package.json');
    console.log(`html-to-markdown-node v${packageJson.version}`);
}
async function readStdin() {
    return new Promise((resolve, reject) => {
        let data = '';
        process.stdin.setEncoding('utf8');
        process.stdin.on('data', chunk => {
            data += chunk;
        });
        process.stdin.on('end', () => {
            resolve(data);
        });
        process.stdin.on('error', reject);
    });
}
async function main() {
    const args = process.argv.slice(2);
    const options = parseArgs(args);
    if (options.help) {
        showHelp();
        process.exit(0);
    }
    if (options.version) {
        showVersion();
        process.exit(0);
    }
    try {
        let html;
        if (options.input) {
            if (!fs.existsSync(options.input)) {
                console.error(`Error: Input file not found: ${options.input}`);
                process.exit(1);
            }
            html = fs.readFileSync(options.input, 'utf8');
        }
        else {
            if (process.stdin.isTTY) {
                console.error('Error: No input provided. Use -i <file> or pipe HTML to stdin.');
                console.error('Run with --help for usage information.');
                process.exit(1);
            }
            html = await readStdin();
        }
        const markdown = (0, index_1.convertString)(html, {
            domain: options.domain
        });
        if (options.output) {
            const outputDir = path.dirname(options.output);
            if (!fs.existsSync(outputDir)) {
                fs.mkdirSync(outputDir, { recursive: true });
            }
            fs.writeFileSync(options.output, markdown, 'utf8');
            console.error(`✓ Converted successfully: ${options.output}`);
        }
        else {
            console.log(markdown);
        }
    }
    catch (error) {
        console.error('Error:', error instanceof Error ? error.message : error);
        process.exit(1);
    }
}
if (require.main === module) {
    main().catch(error => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}
