// @siping/html-to-markdown-node/dist/Converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Converter = void 0;
const jsdom_1 = require("jsdom");
const Register_1 = require("./Register");
const escape_1 = require("./utils/escape");
const dom_1 = require("./utils/dom");
class Converter {
    constructor(options = {}) {
        this.escapeMode = 'smart';
        this.registeredPlugins = [];
        this.register = new Register_1.Register(this);
        if (options.escapeMode) {
            this.escapeMode = options.escapeMode;
        }
        if (options.plugins) {
            options.plugins.forEach(plugin => this.registerPlugin(plugin));
        }
    }
    registerPlugin(plugin) {
        if (!plugin.name) {
            throw new Error('Plugin must have a name');
        }
        this.registeredPlugins.push(plugin.name);
        plugin.init(this);
    }
    convertString(html, options = {}) {
        const dom = new jsdom_1.JSDOM(html);
        const doc = dom.window.document;
        return this.convertDocument(doc, options);
    }
    convertDocument(doc, options) {
        const ctx = this.createContext(options);
        const preHandlers = this.register.preRenderHandlers.getAll();
        preHandlers.forEach(handler => handler(ctx, doc));
        const content = this.renderNode(ctx, doc.body || doc);
        let result = content;
        const postHandlers = this.register.postRenderHandlers.getAll();
        postHandlers.forEach(handler => {
            result = handler(ctx, result);
        });
        return result.trim();
    }
    createContext(options) {
        const ctx = {
            domain: options.domain,
            getTagType: (tagName) => this.register.getTagType(tagName),
            shouldEscape: (char) => this.register.shouldEscape(char),
            convertNode: (node) => this.convertNodeContent(node, ctx)
        };
        return ctx;
    }
    renderNode(ctx, node) {
        if (!node)
            return '';
        if ((0, dom_1.isTextNode)(node)) {
            return this.renderTextNode(ctx, node);
        }
        if (!(0, dom_1.isElement)(node)) {
            if ((0, dom_1.hasChildNodes)(node)) {
                return this.renderChildren(ctx, node);
            }
            return '';
        }
        const handlers = this.register.renderHandlers.getAll();
        for (const handler of handlers) {
            const result = handler(ctx, node);
            if (result !== 'try-next') {
                return result;
            }
        }
        return this.renderChildren(ctx, node);
    }
    convertNodeContent(node, ctx) {
        return this.renderChildren(ctx, node);
    }
    renderTextNode(ctx, node) {
        let text = (0, dom_1.getTextContent)(node);
        const transformers = this.register.textTransformHandlers.getAll();
        transformers.forEach(transformer => {
            text = transformer(ctx, text);
        });
        if (this.escapeMode === 'smart') {
            const charsToEscape = new Set();
            for (const char of text) {
                if (ctx.shouldEscape(char)) {
                    charsToEscape.add(char);
                }
            }
            text = (0, escape_1.escapeMarkdown)(text, charsToEscape);
        }
        return text;
    }
    renderChildren(ctx, node) {
        if (!(0, dom_1.hasChildNodes)(node)) {
            return '';
        }
        const parts = [];
        const children = Array.from(node.childNodes);
        for (const child of children) {
            const content = this.renderNode(ctx, child);
            if (content) {
                parts.push(content);
            }
        }
        return parts.join('');
    }
}
exports.Converter = Converter;
