// @siping/html-to-markdown-node/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TablePlugin = exports.CommonMarkPlugin = exports.BasePlugin = exports.Converter = void 0;
exports.convertString = convertString;
const Converter_1 = require("./Converter");
const base_1 = require("./plugins/base");
const commonmark_1 = require("./plugins/commonmark");
const table_1 = require("./plugins/table");
function convertString(html, options = {}) {
    const converter = new Converter_1.Converter({
        plugins: [
            new base_1.BasePlugin(),
            new commonmark_1.CommonMarkPlugin(),
            new table_1.TablePlugin()
        ]
    });
    return converter.convertString(html, options);
}
var Converter_2 = require("./Converter");
Object.defineProperty(exports, "Converter", { enumerable: true, get: function () { return Converter_2.Converter; } });
var base_2 = require("./plugins/base");
Object.defineProperty(exports, "BasePlugin", { enumerable: true, get: function () { return base_2.BasePlugin; } });
var commonmark_2 = require("./plugins/commonmark");
Object.defineProperty(exports, "CommonMarkPlugin", { enumerable: true, get: function () { return commonmark_2.CommonMarkPlugin; } });
var table_2 = require("./plugins/table");
Object.defineProperty(exports, "TablePlugin", { enumerable: true, get: function () { return table_2.TablePlugin; } });
__exportStar(require("./types"), exports);
__exportStar(require("./Register"), exports);
