// @siping/html-to-markdown-node/dist/Register.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Register = exports.PriorityLate = exports.PriorityStandard = exports.PriorityEarly = void 0;
const prioritized_1 = require("./utils/prioritized");
const dom_1 = require("./utils/dom");
exports.PriorityEarly = 10;
exports.PriorityStandard = 100;
exports.PriorityLate = 1000;
class Register {
    constructor(converter) {
        this.preRenderHandlers = new prioritized_1.PrioritizedList();
        this.renderHandlers = new prioritized_1.PrioritizedList();
        this.postRenderHandlers = new prioritized_1.PrioritizedList();
        this.textTransformHandlers = new prioritized_1.PrioritizedList();
        this.tagTypes = new Map();
        this.escapedChars = new Set();
        this.converter = converter;
    }
    preRenderer(fn, priority = exports.PriorityStandard) {
        this.preRenderHandlers.add(fn, priority);
    }
    renderer(fn, priority = exports.PriorityStandard) {
        this.renderHandlers.add(fn, priority);
    }
    rendererFor(tagName, tagType, renderFn, priority = exports.PriorityStandard) {
        this.tagType(tagName, tagType, priority);
        this.renderer((ctx, node) => {
            if (node.nodeType === 1 && node.tagName.toLowerCase() === tagName) {
                return renderFn(ctx, node);
            }
            return 'try-next';
        }, priority);
    }
    postRenderer(fn, priority = exports.PriorityStandard) {
        this.postRenderHandlers.add(fn, priority);
    }
    textTransformer(fn, priority = exports.PriorityStandard) {
        this.textTransformHandlers.add(fn, priority);
    }
    tagType(tagName, type, priority = exports.PriorityStandard) {
        if (!this.tagTypes.has(tagName)) {
            this.tagTypes.set(tagName, new prioritized_1.PrioritizedList());
        }
        this.tagTypes.get(tagName).add(type, priority);
    }
    getTagType(tagName) {
        const types = this.tagTypes.get(tagName);
        if (types) {
            const all = types.getAll();
            if (all.length > 0) {
                return all[0];
            }
        }
        if ((0, dom_1.isBlockElement)(tagName)) {
            return 'block';
        }
        else if ((0, dom_1.isInlineElement)(tagName)) {
            return 'inline';
        }
        return undefined;
    }
    escapedChar(...chars) {
        chars.forEach(char => this.escapedChars.add(char));
    }
    shouldEscape(char) {
        return this.escapedChars.has(char);
    }
}
exports.Register = Register;
