// news-to-markdown/dist/core/converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewsToMarkdownConverter = void 0;
const html_to_markdown_node_1 = require("@siping/html-to-markdown-node");
const platforms_1 = require("../platforms");
const fetcher_1 = require("./fetcher");
const dual_extractor_1 = require("./dual-extractor");
const image_downloader_1 = require("../utils/image-downloader");
const path_1 = require("path");
class NewsToMarkdownConverter {
    constructor() {
        this.platformRegistry = new platforms_1.PlatformRegistry();
        this.fetcher = new fetcher_1.Fetcher();
    }
    registerPlatform(platform) {
        this.platformRegistry.register(platform);
    }
    async convert(options) {
        const verbose = options.verbose || false;
        this.fetcher = new fetcher_1.Fetcher(verbose);
        if (options.customPlatform) {
            this.platformRegistry.register(options.customPlatform);
        }
        // 先检测平台，以便使用平台特定的 URL 转换
        const platform = options.platform
            ? this.platformRegistry.get(options.platform)
            : this.platformRegistry.detect(options.url);
        // 如果平台支持移动端 URL 转换（用于绕过反爬机制），则转换 URL
        let fetchUrl = options.url;
        if (platform.getMobileUrl) {
            fetchUrl = platform.getMobileUrl(options.url);
            if (verbose && fetchUrl !== options.url) {
                console.error(`[Platform] ${platform.name} 使用移动端 URL: ${fetchUrl}`);
            }
        }
        let html;
        let fetchMethod = 'pre-fetched';
        if (options.html) {
            if (verbose)
                console.error('[Fetch] 使用预抓取 HTML，跳过网络请求');
            html = options.html;
        }
        else {
            if (verbose)
                console.error(`[Fetch] 开始抓取: ${fetchUrl}`);
            const fetchResult = await this.fetcher.fetch(fetchUrl, options.fetchStrategy || 'auto', options.timeout || 30000);
            if (!fetchResult.success) {
                throw new Error(`抓取失败: ${fetchResult.error}`);
            }
            html = fetchResult.html;
            fetchMethod = fetchResult.method;
        }
        if (verbose)
            console.error(`[Platform] 检测到平台: ${platform.name}`);
        if (verbose)
            console.error('[Platform] 平台预处理...');
        const originalHtml = html;
        html = platform.preprocess(html, options.url);
        if (verbose)
            console.error('[Extract] 开始双引擎内容提取...');
        const extractor = new dual_extractor_1.DualExtractor(verbose);
        const extracted = await extractor.extract(html, {
            url: options.url,
            noiseSelectors: options.noiseSelectors || [],
            verbose,
            platform,
            originalHtml,
        });
        if (verbose)
            console.error('[Convert] 转换为 Markdown...');
        let markdown = (0, html_to_markdown_node_1.convertString)(extracted.contentHtml, {
            domain: new URL(options.url).hostname,
        });
        if (platform.postprocess) {
            if (verbose)
                console.error('[Platform] 平台后处理...');
            markdown = platform.postprocess(markdown);
        }
        // 修复代码块内的空行格式
        markdown = this.normalizeCodeBlockSpacing(markdown);
        // 清理尾部参考链接
        markdown = this.cleanTrailingLinks(markdown, verbose);
        // 压缩连续空行（3+ 个空行压缩为 2 个）
        markdown = this.compressEmptyLines(markdown, verbose);
        // 移除正文中与标题和作者重复的内容
        markdown = this.removeDuplicateTitleAuthor(markdown, extracted.title, extracted.author, verbose);
        if (options.includeMetadata !== false) {
            markdown = this.addMetadata(markdown, extracted, options.url);
        }
        // 下载图片到本地（如果启用）
        let localCoverPath;
        let downloadedImagesCount = 0;
        if (options.downloadImages && options.outputDir) {
            if (verbose)
                console.error('[Download] 开始下载图片到本地...');
            const imagesDir = (0, path_1.join)(options.outputDir, 'images');
            const imagesToDownload = [];
            const markdownImageMap = new Map();
            // 收集需要下载的图片 URL
            if (extracted.coverImage && extracted.coverImage.startsWith('http')) {
                imagesToDownload.push(extracted.coverImage);
            }
            // 收集正文中的远程图片
            for (const img of extracted.images) {
                if (img.startsWith('http') && !imagesToDownload.includes(img)) {
                    imagesToDownload.push(img);
                }
            }
            if (imagesToDownload.length > 0) {
                // 下载封面图
                if (extracted.coverImage && extracted.coverImage.startsWith('http')) {
                    const coverResult = await (0, image_downloader_1.downloadImage)({
                        url: extracted.coverImage,
                        outputDir: imagesDir,
                        filename: 'cover' + this.getExtensionFromUrl(extracted.coverImage),
                    });
                    if (coverResult.success && coverResult.localPath) {
                        localCoverPath = coverResult.localPath;
                        markdownImageMap.set(extracted.coverImage, coverResult.localPath);
                        downloadedImagesCount++;
                        if (verbose)
                            console.error(`[Download] ✓ 封面图下载成功: ${localCoverPath}`);
                    }
                    else {
                        if (verbose)
                            console.error(`[Download] ✗ 封面图下载失败: ${coverResult.error}`);
                    }
                }
                // 下载正文图片（只下载 markdown 中仍然存在的图片，排除被 postprocess 移除的）
                // 注意：需要用 URL 的基础部分匹配，因为 postprocess 可能清理了 #imgIndex 等后缀
                const contentImages = extracted.images.filter((img) => {
                    if (!img.startsWith('http') || img === extracted.coverImage)
                        return false;
                    // 提取 URL 基础部分（去掉 #hash 和部分 query 参数）
                    const baseUrl = img.split('#')[0].split('?')[0];
                    return markdown.includes(baseUrl);
                });
                if (contentImages.length > 0) {
                    const urlToLocalPath = await (0, image_downloader_1.downloadImages)(contentImages, imagesDir, verbose);
                    downloadedImagesCount += urlToLocalPath.size;
                    for (const [url, localPath] of urlToLocalPath.entries()) {
                        markdownImageMap.set(url, localPath);
                    }
                    if (verbose)
                        console.error(`[Download] ✓ 下载了 ${urlToLocalPath.size} 张正文图片`);
                }
                // 替换 Markdown 中的图片 URL（包含正文图片和正文里重复出现的封面图）
                if (markdownImageMap.size > 0) {
                    markdown = (0, image_downloader_1.replaceImageUrls)(markdown, markdownImageMap, imagesDir);
                }
                // 更新 frontmatter 中的 cover 路径
                if (localCoverPath) {
                    const relativeCoverPath = './images/' + localCoverPath.split('/').pop();
                    markdown = markdown.replace(/^cover: "[^"]+"$/m, `cover: "${relativeCoverPath}"`);
                }
            }
            if (verbose)
                console.error(`[Download] 总共下载了 ${downloadedImagesCount} 张图片`);
        }
        return {
            markdown,
            metadata: {
                title: extracted.title,
                author: extracted.author,
                publishTime: extracted.publishTime,
                platform: platform.name,
                imageCount: extracted.images.length,
                contentLength: markdown.length,
                fetchMethod,
                coverImage: extracted.coverImage || null,
                localCoverPath,
                downloadedImages: downloadedImagesCount,
            },
            html: options.verbose ? html : undefined,
        };
    }
    addMetadata(markdown, extracted, url) {
        const parts = [];
        // 添加 frontmatter
        const frontmatter = ['---'];
        if (extracted.title) {
            // 转义 YAML 中的双引号，保留原始引号
            const escapedTitle = extracted.title.replace(/"/g, '\\"');
            frontmatter.push(`title: "${escapedTitle}"`);
        }
        if (extracted.author) {
            const escapedAuthor = extracted.author.replace(/"/g, '\\"');
            frontmatter.push(`author: "${escapedAuthor}"`);
        }
        if (extracted.publishTime)
            frontmatter.push(`date: "${extracted.publishTime}"`);
        if (extracted.coverImage)
            frontmatter.push(`cover: "${extracted.coverImage}"`);
        frontmatter.push('---\n');
        parts.push(frontmatter.join('\n'));
        // 不再添加重复的可读元数据，避免与 frontmatter 冲突
        // frontmatter 中已经包含了所有必要的元数据信息
        // 移除正文开头与封面相同的重复图片
        let cleanedMarkdown = markdown;
        if (extracted.coverImage) {
            cleanedMarkdown = this.removeDuplicateCoverImage(markdown, extracted.coverImage);
        }
        parts.push(cleanedMarkdown);
        // 图片已经在正文中，不需要单独列出
        // if (extracted.images && extracted.images.length > 0) {
        //   parts.push('\n## 图片列表\n');
        //   extracted.images.forEach((img: string) => {
        //     parts.push(`- ![](${img})\n`);
        //   });
        // }
        return parts.join('\n');
    }
    /**
     * 移除正文开头与封面相同的重复图片
     * 只移除正文最开始位置的图片（在第一个标题或段落之前）
     */
    removeDuplicateCoverImage(markdown, coverImage) {
        const lines = markdown.split('\n');
        let firstImageIndex = -1;
        let firstContentIndex = -1;
        // 查找第一张图片和第一个内容的位置
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i].trim();
            // 跳过空行
            if (!line)
                continue;
            // 检查是否是图片
            const imgMatch = line.match(/^!\[.*?\]\((.*?)\)/);
            if (imgMatch && firstImageIndex === -1) {
                const imgUrl = imgMatch[1];
                // 比较图片 URL（忽略查询参数的差异）
                const coverUrlBase = coverImage.split('?')[0];
                const imgUrlBase = imgUrl.split('?')[0];
                if (coverUrlBase === imgUrlBase || imgUrl === coverImage) {
                    firstImageIndex = i;
                    continue;
                }
            }
            // 检查是否是实际内容（标题、段落等）
            if (line.startsWith('#') || (line.length > 0 && !line.startsWith('!'))) {
                firstContentIndex = i;
                break;
            }
        }
        // 如果第一张图片在第一个内容之前，且与封面相同，则移除它
        if (firstImageIndex !== -1 && (firstContentIndex === -1 || firstImageIndex < firstContentIndex)) {
            lines.splice(firstImageIndex, 1);
            return lines.join('\n');
        }
        return markdown;
    }
    /**
     * 移除正文开头与标题和作者重复的内容
     */
    removeDuplicateTitleAuthor(markdown, title, author, verbose) {
        if (!title && !author)
            return markdown;
        const lines = markdown.split('\n');
        const resultLines = [];
        let removedCount = 0;
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            const trimmedLine = line.trim();
            // 保留空行（但压缩连续空行），避免段落在渲染时被合并
            if (!trimmedLine) {
                if (resultLines.length > 0 && resultLines[resultLines.length - 1].trim() !== '') {
                    resultLines.push('');
                }
                continue;
            }
            // 跳过空标题
            if (trimmedLine === '#' || trimmedLine === '##') {
                continue;
            }
            // 检查是否与标题重复（完全匹配或去掉 # 后匹配）
            if (title && (trimmedLine === title || trimmedLine.replace(/^#+\s*/, '') === title)) {
                removedCount++;
                continue;
            }
            // 检查是否与作者重复
            if (author && trimmedLine === author) {
                removedCount++;
                continue;
            }
            // 检查是否是作者名后面跟着空格或特殊字符
            if (author && trimmedLine.startsWith(author) && trimmedLine.replace(author, '').trim().match(/^[\s;:,，。、]*$/)) {
                removedCount++;
                continue;
            }
            resultLines.push(line);
        }
        if (verbose && removedCount > 0) {
            console.error(`[Cleaner] 移除了 ${removedCount} 行重复的标题/作者信息`);
        }
        return resultLines.join('\n');
    }
    /**
     * 从 URL 获取扩展名
     */
    getExtensionFromUrl(url) {
        const urlWithoutQuery = url.split('?')[0];
        const parts = urlWithoutQuery.split('/');
        const filename = parts[parts.length - 1];
        const match = filename.match(/\.(jpg|jpeg|png|gif|webp|bmp)$/i);
        return match ? match[0] : '.jpg';
    }
    /**
     * 确保代码块前后有正确的空行格式
     * 移除代码块开始后和结束前的空行
     */
    normalizeCodeBlockSpacing(markdown) {
        const lines = markdown.split('\n');
        const normalized = [];
        let inCodeBlock = false;
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            const trimmed = line.trim();
            if (trimmed.startsWith('```')) {
                if (!inCodeBlock) {
                    // 代码块开始
                    // 确保代码块前有空行（如果前面不是空行）
                    if (normalized.length > 0 && normalized[normalized.length - 1].trim() !== '') {
                        normalized.push('');
                    }
                    normalized.push(line);
                    inCodeBlock = true;
                    // 跳过代码块开始后的空行
                    while (i + 1 < lines.length && lines[i + 1].trim() === '') {
                        i++;
                    }
                }
                else {
                    // 代码块结束
                    // 移除代码块结束前的空行
                    while (normalized.length > 0 && normalized[normalized.length - 1].trim() === '') {
                        normalized.pop();
                    }
                    normalized.push(line);
                    inCodeBlock = false;
                }
            }
            else {
                normalized.push(line);
            }
        }
        return normalized.join('\n');
    }
    /**
     * 压缩连续的空行
     * 将 3 个或更多连续空行压缩为 2 个空行
     * 这样可以防止 wenyan-md 生成过多的空 <p> 标签
     */
    compressEmptyLines(markdown, verbose) {
        const before = markdown.match(/\n{3,}/g)?.length || 0;
        const compressed = markdown.replace(/\n{3,}/g, '\n\n');
        const after = compressed.match(/\n{3,}/g)?.length || 0;
        if (verbose && before > 0) {
            console.error(`[Cleaner] 压缩了 ${before} 处连续空行`);
        }
        return compressed;
    }
    /**
     * 清理 Markdown 末尾的参考链接区域
     * 匹配形如 "参考资料:", "来源:", "参考链接:", "Reference:" 等段落及其中的链接
     */
    cleanTrailingLinks(markdown, verbose) {
        const lines = markdown.split('\n');
        const refPatterns = [
            /^参考资料?[:：]/i,
            /^来源?[:：]/i,
            /^参考链接?[:：]/i,
            /^reference?s?[:：]/i,
            /^ref\.?[:：]/i,
            /^\[?\d+\]?\s*参考/i,
            /^延伸阅读/i,
            /^相关阅读/i,
            /^相关链接?[:：]/i,
            /^external links?[:：]/i,
        ];
        // 找到参考链接区的起始行
        let refStart = -1;
        for (let i = lines.length - 1; i >= 0; i--) {
            const line = lines[i].trim();
            if (!line)
                continue;
            // 如果是纯链接行（行内只有 Markdown 链接或 URL），可能是参考链接的一部分
            const isLinkOnly = /^\[.+\]\(.+\)$/.test(line) || /^https?:\/\/.+/i.test(line);
            // 如果是参考标题行
            for (const pattern of refPatterns) {
                if (pattern.test(line)) {
                    refStart = i;
                    break;
                }
            }
            // 如果遇到空行且还没找到参考标题，说明参考区已结束
            if (refStart === -1 && line === '' && i < lines.length - 3) {
                // 检查这之前是否有一系列链接行
                let linkCount = 0;
                for (let j = i - 1; j >= 0; j--) {
                    const prev = lines[j].trim();
                    if (/^\[.+\]\(.+\)$/.test(prev) || /^https?:\/\/.+/i.test(prev)) {
                        linkCount++;
                    }
                    else if (prev) {
                        break;
                    }
                }
                if (linkCount >= 2) {
                    // 找到了多个尾部链接，但没有明确的标题，删除它们
                    refStart = i - linkCount + 1;
                }
            }
            if (refStart !== -1)
                break;
        }
        if (refStart !== -1) {
            if (verbose)
                console.error(`[Cleaner] 发现尾部参考链接，清理 ${lines.length - refStart} 行`);
            return lines.slice(0, refStart).join('\n').trimEnd();
        }
        return markdown;
    }
}
exports.NewsToMarkdownConverter = NewsToMarkdownConverter;
