// news-to-markdown/dist/core/dual-extractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DualExtractor = void 0;
const jsdom_1 = require("jsdom");
const readability_1 = require("@mozilla/readability");
const news_extractor_node_1 = require("news-extractor-node");
const cheerio_1 = require("cheerio");
class DualExtractor {
    constructor(verbose = false) {
        this.verbose = verbose;
    }
    async extract(html, options) {
        const { url, noiseSelectors = [], preferReadability = false } = options;
        if (this.verbose)
            console.error('[DualExtractor] 开始双引擎内容提取...');
        // 0. 对于微信文章，优先直接提取 #js_content（Readability 经常截断微信内容）
        const isWechat = url.includes('mp.weixin.qq.com');
        const wechatDirectContent = isWechat ? this.extractWechatDirectContent(html) : null;
        // 1. 使用 Readability 提取主体内容和结构
        const readabilityResult = this.extractWithReadability(html, url);
        // 2. 使用 NewsExtractor 提取元数据
        const newsResult = this.extractWithNewsExtractor(html, url, noiseSelectors);
        // 3. 决定使用哪个引擎的内容
        let finalContent;
        let finalContentHtml;
        const readabilityLength = readabilityResult.textContent?.length || 0;
        const readabilityNonWhitespace = (readabilityResult.textContent || '').replace(/\s/g, '').length;
        const newsLength = newsResult.content?.length || 0;
        const wechatDirectLength = wechatDirectContent?.textLength || 0;
        // 如果 NewsExtractor 返回的内容异常大（超过 100KB），可能是整个 HTML，应该使用 Readability
        const newsContentTooLarge = newsLength > 100000;
        // 对于微信文章，如果直接提取有内容，优先使用直接提取（Readability 对微信 section 结构提取会包含大量空白）
        // 用非空白字符数对比，避免 Readability 因空白多而错误胜出
        if (wechatDirectContent && wechatDirectLength > readabilityNonWhitespace * 1.1) {
            if (this.verbose) {
                console.error(`[DualExtractor] 微信直接提取内容更完整 (${wechatDirectLength} vs ${readabilityLength} 字符)，使用直接提取`);
            }
            finalContentHtml = wechatDirectContent.html;
            finalContent = wechatDirectContent.text;
        }
        else if (preferReadability || !newsResult.contentHtml || newsContentTooLarge) {
            if (this.verbose) {
                if (newsContentTooLarge) {
                    console.error(`[DualExtractor] NewsExtractor 内容异常大 (${newsLength} 字符)，使用 Readability`);
                }
                else {
                    console.error('[DualExtractor] 使用 Readability 提取的内容');
                }
            }
            finalContentHtml = readabilityResult.content || '';
            finalContent = readabilityResult.textContent || '';
        }
        else {
            // 比较两个结果，选择更好的
            if (readabilityLength > newsLength * 1.2) {
                // Readability 提取的内容明显更多（可能保留了图片）
                if (this.verbose)
                    console.error('[DualExtractor] Readability 内容更完整，使用 Readability');
                finalContentHtml = readabilityResult.content || '';
                finalContent = readabilityResult.textContent || '';
            }
            else {
                if (this.verbose)
                    console.error('[DualExtractor] 使用 NewsExtractor 提取的内容');
                finalContentHtml = newsResult.contentHtml;
                finalContent = newsResult.content || '';
            }
        }
        // 如果两个引擎都提取失败，使用预处理后的原始 HTML（适用于纯图片文章）
        if (!finalContentHtml && html) {
            if (this.verbose)
                console.error('[DualExtractor] 两个引擎都提取失败，使用预处理后的原始 HTML');
            // 从原始 HTML 中提取图片，构建简单的图片列表 HTML
            const $ = (0, cheerio_1.load)(html);
            const imgHtmlParts = [];
            $('img').each((i, el) => {
                const src = $(el).attr('src') || '';
                if (src && !src.startsWith('data:image')) {
                    imgHtmlParts.push(`<p><img src="${src}" alt="图片"></p>`);
                }
            });
            finalContentHtml = imgHtmlParts.join('\n');
            finalContent = '';
        }
        // 4. 提取所有图片（同时清理小图标）
        // 优先从预处理后的 HTML 提取图片（Readability 可能丢失图片）
        let { images, cleanedHtml } = this.extractAllImages(finalContentHtml);
        // 如果从内容中没有提取到图片，尝试从预处理后的完整 HTML 中提取
        if (images.length === 0) {
            const htmlImages = this.extractAllImages(html);
            if (htmlImages.images.length > 0) {
                images = htmlImages.images;
                // 将图片添加到内容 HTML 中
                const imgHtml = images.map(src => `<p><img src="${src}" alt=""></p>`).join('\n');
                finalContentHtml = cleanedHtml + '\n' + imgHtml;
                if (this.verbose)
                    console.error(`[DualExtractor] 从预处理 HTML 中补充提取了 ${images.length} 张图片`);
            }
        }
        finalContentHtml = cleanedHtml;
        // 5. Select cover image
        let coverImage = this.selectCoverImage(html, images);
        // 6. Merge metadata (priority: platform metaSelectors > Readability > NewsExtractor)
        // Use original HTML (before preprocessing) to extract meta info, avoiding meta tag removal
        const htmlForMeta = options.originalHtml || html;
        const platformMeta = options.platform?.metaSelectors
            ? this.extractWithSelectors(htmlForMeta, options.platform.metaSelectors)
            : {};
        // For mobile pages, Readability is usually more reliable than NewsExtractor
        let title = platformMeta.title || readabilityResult.title || newsResult.title || undefined;
        // 清理标题中的平台后缀
        if (title) {
            title = title.replace(/\s*[-–—]\s*(今日头条|头条|Toutiao).*$/i, '').trim();
        }
        // 过滤无效的作者信息
        const rawAuthor = platformMeta.author || readabilityResult.byline || newsResult.author || undefined;
        const author = this.validateAuthor(rawAuthor);
        const publishTime = platformMeta.time || readabilityResult.publishedTime || newsResult.publishTime || undefined;
        // 7. 如果没有封面图且有标题，生成动态封面
        if (!coverImage && title) {
            coverImage = this.generateDynamicCover(title);
            if (this.verbose)
                console.error('[DualExtractor] 已生成动态封面:', coverImage);
        }
        return {
            title,
            author,
            publishTime,
            content: finalContent,
            contentHtml: finalContentHtml,
            images,
            coverImage,
            textLength: finalContent.length,
            excerpt: readabilityResult.excerpt || undefined,
        };
    }
    extractWithReadability(html, url) {
        try {
            const dom = new jsdom_1.JSDOM(html, { url });
            const reader = new readability_1.Readability(dom.window.document, {
                charThreshold: 100,
                keepClasses: false,
            });
            const article = reader.parse();
            if (!article) {
                if (this.verbose)
                    console.error('[DualExtractor] Readability 提取失败');
                return {
                    title: '',
                    content: '',
                    textContent: '',
                    byline: '',
                    excerpt: '',
                    publishedTime: '',
                };
            }
            if (this.verbose) {
                console.error(`[DualExtractor] Readability 提取成功: ${article.textContent?.length || 0} 字符`);
            }
            return article;
        }
        catch (error) {
            if (this.verbose)
                console.error('[DualExtractor] Readability 错误:', error);
            return {
                title: '',
                content: '',
                textContent: '',
                byline: '',
                excerpt: '',
                publishedTime: '',
            };
        }
    }
    /**
     * 直接从微信文章的 #js_content 提取内容
     * Readability 经常截断微信文章内容，这个方法作为备用
     */
    extractWechatDirectContent(html) {
        try {
            const $ = (0, cheerio_1.load)(html);
            const $content = $('#js_content');
            if (!$content.length) {
                return null;
            }
            const contentHtml = $content.html() || '';
            const contentText = $content.text().replace(/\s+/g, ' ').trim();
            if (this.verbose) {
                console.error(`[DualExtractor] 微信直接提取: ${contentText.length} 字符`);
            }
            return {
                html: contentHtml,
                text: contentText,
                textLength: contentText.length,
            };
        }
        catch (error) {
            if (this.verbose)
                console.error('[DualExtractor] 微信直接提取错误:', error);
            return null;
        }
    }
    extractWithNewsExtractor(html, url, noiseSelectors) {
        try {
            const extractor = new news_extractor_node_1.NewsExtractor();
            const result = extractor.extract(html, {
                url,
                noiseSelectors,
            });
            if (this.verbose) {
                console.error(`[DualExtractor] NewsExtractor 提取成功: ${result.content?.length || 0} 字符`);
            }
            return result;
        }
        catch (error) {
            if (this.verbose)
                console.error('[DualExtractor] NewsExtractor 错误:', error);
            return {
                title: '',
                content: '',
                contentHtml: '',
                author: '',
                publishTime: '',
                description: '',
                images: [],
            };
        }
    }
    /**
     * 判断图片是否为小图标/表情包
     * 微信文章中的表情包通常：
     * 1. URL 包含特定关键词
     * 2. 尺寸很小（< 100px）
     */
    isSmallEmojiImage(img, src) {
        // URL 特征词检测
        const emojiPatterns = [
            'emotion',
            'sticker',
            'wxm3',
            'wxm4',
            '/emoji/',
            '/emoticon/',
            '/sticker/',
            'data-src', // 延迟加载的表情包
        ];
        const isEmojiUrl = emojiPatterns.some(p => src.toLowerCase().includes(p));
        // 微信小尺寸参数检测 (wxfrom=18 是缩略图, wxfrom=13 是移动端正常参数不过滤)
        const isSmallWx = src.includes('wxfrom=18');
        // CSS 尺寸检测（只匹配 px，避免把 width:100% 误判为 100px）
        const style = img.attr('style') || '';
        const widthMatch = style.match(/width:\s*(\d+)px/);
        const heightMatch = style.match(/height:\s*(\d+)px/);
        const width = widthMatch ? parseInt(widthMatch[1]) : 0;
        const height = heightMatch ? parseInt(heightMatch[1]) : 0;
        // 如果尺寸小于 120px，认为是小图标
        const isSmall = (width > 0 && width < 120) || (height > 0 && height < 120);
        // 注意：__bg_gif 是微信懒加载通用 class，不作为过滤条件
        return isEmojiUrl || isSmallWx || isSmall;
    }
    extractAllImages(html) {
        try {
            const $ = (0, cheerio_1.load)(html);
            const images = [];
            // 先过滤并移除小图标
            $('img').each((i, el) => {
                const $img = $(el);
                const src = $img.attr('src') || $img.attr('data-src');
                if (!src || src.startsWith('data:image') || images.includes(src)) {
                    return;
                }
                // 注意：__bg_gif 是微信懒加载通用 class，不过滤（GIF 动图为正文内容图）
                // 过滤小图标/表情包
                if (this.isSmallEmojiImage($img, src)) {
                    $img.remove();
                    return;
                }
                images.push(src);
            });
            return {
                images,
                cleanedHtml: $.html()
            };
        }
        catch (error) {
            if (this.verbose)
                console.error('[DualExtractor] 图片提取错误:', error);
            return { images: [], cleanedHtml: html };
        }
    }
    /**
     * 使用平台级 CSS 选择器提取 meta 信息，优先级最高
     */
    extractWithSelectors(html, selectors) {
        const result = {};
        try {
            const $ = (0, cheerio_1.load)(html);
            if (selectors.title) {
                const titleText = this.selectText($, selectors.title);
                if (this.verbose)
                    console.error(`[DualExtractor] selectText(title) 返回: "${titleText}"`);
                result.title = titleText || undefined;
            }
            if (selectors.author) {
                const authorText = this.selectText($, selectors.author);
                if (this.verbose)
                    console.error(`[DualExtractor] selectText(author) 返回: "${authorText}"`);
                result.author = authorText || undefined;
            }
            if (selectors.time) {
                const timeText = this.selectText($, selectors.time);
                if (this.verbose)
                    console.error(`[DualExtractor] selectText(time) 返回: "${timeText}"`);
                result.time = timeText || undefined;
            }
            if (this.verbose) {
                console.error(`[DualExtractor] 平台级 meta 提取:`, result);
            }
        }
        catch (error) {
            if (this.verbose)
                console.error('[DualExtractor] metaSelectors 提取错误:', error);
        }
        return result;
    }
    selectText($, selector) {
        // 支持多个选择器（逗号分隔）
        const selectors = selector.split(',').map(s => s.trim());
        for (const sel of selectors) {
            let text = '';
            // meta 标签特殊处理
            if (sel.startsWith('meta[')) {
                const $el = $(sel);
                text = $el.attr('content') || '';
            }
            else {
                // DOM 选择器 - 只取第一个匹配的元素
                const $el = $(sel).first();
                if ($el.length === 0)
                    continue;
                if ($el.is('time')) {
                    text = $el.attr('datetime') || $el.text().trim();
                }
                else {
                    text = $el.text().trim();
                }
            }
            // 清理和验证提取的文本
            text = text.trim();
            // 过滤无效的作者/标题（常见的错误提取）
            const invalidPatterns = [
                /^following the instructions/i,
                /^click here/i,
                /^read more/i,
                /^undefined$/i,
                /^null$/i,
                /^\s*$/,
                /^https?:\/\//i, // URL 不应该是标题或作者
            ];
            const isInvalid = invalidPatterns.some(pattern => pattern.test(text));
            if (!isInvalid && text.length > 0 && text.length < 200) {
                // 返回第一个有效的结果
                return text;
            }
        }
        return '';
    }
    validateAuthor(author) {
        if (!author)
            return undefined;
        const trimmed = author.trim();
        // 过滤无效的作者名称
        const invalidPatterns = [
            /^following the instructions/i,
            /^click here/i,
            /^read more/i,
            /^undefined$/i,
            /^null$/i,
            /^\s*$/,
            /^https?:\/\//i,
            /^www\./i,
            /^by\s*$/i,
            /^作者\s*$/i,
            /^来源\s*$/i,
        ];
        for (const pattern of invalidPatterns) {
            if (pattern.test(trimmed)) {
                return undefined;
            }
        }
        // 长度验证（作者名通常不会太长）
        if (trimmed.length > 50) {
            return undefined;
        }
        return trimmed;
    }
    selectCoverImage(html, images) {
        try {
            const $ = (0, cheerio_1.load)(html);
            // 1. 优先从 Open Graph meta 标签获取
            const ogImage = $('meta[property="og:image"]').attr('content');
            if (ogImage && !ogImage.startsWith('data:image')) {
                // 检查 og:image 是否是装饰性图片
                const isDecorativeOgImage = this.isDecorativeImage($, ogImage);
                if (isDecorativeOgImage) {
                    if (this.verbose)
                        console.error('[DualExtractor] og:image 是装饰性图片，跳过，使用正文第一张图片');
                    // 使用正文第一张图片作为封面
                    if (images.length > 0) {
                        return images[0];
                    }
                }
                else {
                    if (this.verbose)
                        console.error('[DualExtractor] 使用 og:image 作为封面');
                    return ogImage;
                }
            }
            // 2. 尝试从 Twitter Card meta 标签获取
            const twitterImage = $('meta[name="twitter:image"]').attr('content');
            if (twitterImage && !twitterImage.startsWith('data:image')) {
                if (this.verbose)
                    console.error('[DualExtractor] 使用 twitter:image 作为封面');
                return twitterImage;
            }
            // 3. 使用文章中第一张有效图片（默认行为）
            if (images.length > 0) {
                if (this.verbose)
                    console.error('[DualExtractor] 使用文章中第一张图片作为封面');
                return images[0];
            }
            // 4. 如果没有图片，返回 undefined，由调用方决定是否生成动态封面
            if (this.verbose)
                console.error('[DualExtractor] 文章中未找到图片，需要动态生成封面');
            return undefined;
        }
        catch (error) {
            if (this.verbose)
                console.error('[DualExtractor] 封面图选择错误:', error);
            return images.length > 0 ? images[0] : undefined;
        }
    }
    /**
     * 检查图片是否是装饰性图片
     */
    isDecorativeImage($, imageUrl) {
        // 提取 URL 的关键部分用于匹配
        const urlBase = imageUrl.split('?')[0].split('/').slice(-2).join('/');
        if (this.verbose) {
            console.error(`[DualExtractor] 检查封面图是否是装饰图: ${urlBase}`);
        }
        // 在 HTML 中查找所有匹配的 img 标签
        const $imgs = $('img').filter((_, el) => {
            const src = $(el).attr('src') || '';
            const dataSrc = $(el).attr('data-src') || '';
            return src.includes(urlBase) || dataSrc.includes(urlBase);
        });
        if (this.verbose) {
            console.error(`[DualExtractor] 找到 ${$imgs.length} 个匹配的 img 标签`);
        }
        if ($imgs.length === 0) {
            // 找不到对应的 img 标签，假设不是装饰图
            return false;
        }
        // 检查第一个匹配的图片
        const $img = $imgs.first();
        const imgClass = $img.attr('class') || '';
        const style = $img.attr('style') || '';
        if (this.verbose) {
            console.error(`[DualExtractor] 图片 class: ${imgClass}, style width: ${style.match(/width:\s*(\d+)px/)?.[1] || 'N/A'}`);
        }
        // 检查是否是 __bg_gif
        if (imgClass.includes('__bg_gif')) {
            return true;
        }
        // 检查尺寸是否小于等于 60px
        const widthMatch = style.match(/width:\s*(\d+)px/);
        if (widthMatch && parseInt(widthMatch[1]) <= 60) {
            return true;
        }
        return false;
    }
    /**
     * 生成动态封面图 URL
     * 使用第三方服务生成基于标题的封面图
     */
    generateDynamicCover(title) {
        // 使用 DiceBear API 生成抽象图案作为封面
        // 或者使用 placeholder.com 等服务
        const encodedTitle = encodeURIComponent(title.substring(0, 50));
        // 方案1: 使用 placeholder.com 生成纯色背景 + 文字
        // return `https://via.placeholder.com/1200x630/4A90E2/FFFFFF?text=${encodedTitle}`;
        // 方案2: 使用 DiceBear 生成抽象图案（更美观）
        const seed = title.split('').reduce((acc, char) => acc + char.charCodeAt(0), 0);
        return `https://api.dicebear.com/7.x/shapes/svg?seed=${seed}&backgroundColor=4A90E2,E94B3C,6C5CE7,00B894&size=1200`;
    }
}
exports.DualExtractor = DualExtractor;
