// news-to-markdown/dist/core/fetcher.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fetcher = void 0;
const child_process_1 = require("child_process");
class Fetcher {
    constructor(verbose = false) {
        this.verbose = verbose;
    }
    async fetch(url, strategy = 'auto', timeout = 30000) {
        // 头条页面（PC 和移动端）需要 JavaScript 渲染
        // 策略：先手机版，无图片则桌面版，都无图片则返回手机版
        if (strategy === 'auto' && (url.includes('toutiao.com/article/') || url.includes('m.toutiao.com'))) {
            if (this.verbose)
                console.error('[Fetch] 检测到头条页面，使用 Playwright 渲染');
            return this.fetchToutiaoWithFallback(url, timeout);
        }
        // 知乎页面需要 JavaScript 渲染，且需要真实 Chrome 指纹绕过 zse-ck 反爬
        if (strategy === 'auto' && url.includes('zhihu.com')) {
            if (this.verbose)
                console.error('[Fetch] 检测到知乎页面，优先使用真实 Chrome 绕过反爬');
            return this.fetchWithRealChrome(url, timeout);
        }
        // 掘金页面需要 JavaScript 渲染
        if (strategy === 'auto' && url.includes('juejin.cn')) {
            if (this.verbose)
                console.error('[Fetch] 检测到掘金页面，使用 Playwright 渲染');
            return this.fetchWithPlaywright(url, timeout);
        }
        // CSDN 页面需要 JavaScript 渲染
        if (strategy === 'auto' && url.includes('csdn.net')) {
            if (this.verbose)
                console.error('[Fetch] 检测到 CSDN 页面，使用 Playwright 渲染');
            return this.fetchWithPlaywright(url, timeout);
        }
        // 开源中国页面需要 JavaScript 渲染
        if (strategy === 'auto' && url.includes('oschina.net')) {
            if (this.verbose)
                console.error('[Fetch] 检测到开源中国页面，使用 Playwright 渲染');
            return this.fetchWithPlaywright(url, timeout);
        }
        // 微信文章页面需要 JavaScript 渲染（绕过反爬验证）
        if (strategy === 'auto' && url.includes('mp.weixin.qq.com')) {
            if (this.verbose)
                console.error('[Fetch] 检测到微信文章页面，使用 Playwright 渲染');
            return this.fetchWithPlaywright(url, timeout);
        }
        // 36kr 页面需要 JavaScript 渲染，使用移动端 URL 绕过验证码
        if (strategy === 'auto' && url.includes('36kr.com')) {
            // 将 36kr.com 转换为 m.36kr.com
            const mobileUrl = url.replace('://36kr.com', '://m.36kr.com').replace('://www.36kr.com', '://m.36kr.com');
            if (this.verbose)
                console.error('[Fetch] 检测到 36kr 页面，使用移动端 URL:', mobileUrl);
            return this.fetchWithPlaywright(mobileUrl, timeout, 'mobile');
        }
        if (strategy === 'auto') {
            return this.fetchAuto(url, timeout);
        }
        switch (strategy) {
            case 'curl':
                return this.fetchWithCurl(url, timeout);
            case 'wget':
                return this.fetchWithWget(url, timeout);
            case 'playwright':
                return this.fetchWithPlaywright(url, timeout);
            default:
                throw new Error(`Unknown fetch strategy: ${strategy}`);
        }
    }
    async fetchAuto(url, timeout) {
        if (this.verbose)
            console.error('[Layer 1/3] 尝试 curl...');
        const curlResult = await this.fetchWithCurl(url, timeout);
        if (curlResult.success) {
            if (this.verbose)
                console.error(`[Layer 1/3] ✅ curl 成功 (${curlResult.html.length} bytes)`);
            return curlResult;
        }
        if (this.verbose)
            console.error(`[Layer 1/3] ❌ curl 失败: ${curlResult.error}`);
        if (this.verbose)
            console.error('[Layer 2/3] 尝试 wget...');
        const wgetResult = await this.fetchWithWget(url, timeout);
        if (wgetResult.success) {
            if (this.verbose)
                console.error(`[Layer 2/3] ✅ wget 成功 (${wgetResult.html.length} bytes)`);
            return wgetResult;
        }
        if (this.verbose)
            console.error(`[Layer 2/3] ❌ wget 失败: ${wgetResult.error}`);
        if (this.verbose)
            console.error('[Layer 3/3] 尝试 Playwright...');
        const playwrightResult = await this.fetchWithPlaywright(url, timeout);
        if (playwrightResult.success) {
            if (this.verbose)
                console.error(`[Layer 3/3] ✅ Playwright 成功 (${playwrightResult.html.length} bytes)`);
            return playwrightResult;
        }
        if (this.verbose)
            console.error(`[Layer 3/3] ❌ Playwright 失败: ${playwrightResult.error}`);
        throw new Error('所有抓取方式均失败: curl/wget/Playwright 全部失败');
    }
    async fetchWithCurl(url, timeout) {
        try {
            const curlCmd = [
                'curl',
                '-s',
                '-L',
                '-A', '"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"',
                '--max-time', Math.floor(timeout / 1000).toString(),
                '--compressed',
                `"${url}"`
            ].join(' ');
            const html = (0, child_process_1.execSync)(curlCmd, { encoding: 'utf-8', timeout });
            if (html && html.length > 1000) {
                return { html, method: 'curl', success: true };
            }
            return { html: '', method: 'curl', success: false, error: 'HTML too short' };
        }
        catch (error) {
            return {
                html: '',
                method: 'curl',
                success: false,
                error: error instanceof Error ? error.message : String(error)
            };
        }
    }
    async fetchWithWget(url, timeout) {
        try {
            const wgetCmd = [
                'wget',
                '-q',
                '-O', '-',
                '-U', '"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36"',
                `--timeout=${Math.floor(timeout / 1000)}`,
                '--compression=auto',
                `"${url}"`
            ].join(' ');
            const html = (0, child_process_1.execSync)(wgetCmd, { encoding: 'utf-8', timeout });
            if (html && html.length > 1000) {
                return { html, method: 'wget', success: true };
            }
            return { html: '', method: 'wget', success: false, error: 'HTML too short' };
        }
        catch (error) {
            return {
                html: '',
                method: 'wget',
                success: false,
                error: error instanceof Error ? error.message : String(error)
            };
        }
    }
    /**
     * 头条页面抓取策略：先手机版，无图片则桌面版，都无图片则返回手机版
     */
    async fetchToutiaoWithFallback(url, timeout) {
        // 1. 先尝试手机版
        if (this.verbose)
            console.error('[Fetch] 头条策略：先尝试手机版...');
        const mobileResult = await this.fetchWithPlaywright(url, timeout, 'mobile');
        if (!mobileResult.success) {
            if (this.verbose)
                console.error('[Fetch] 手机版抓取失败，尝试桌面版...');
            return this.fetchWithPlaywright(url, timeout, 'desktop');
        }
        // 2. 检查手机版是否有图片（在 article 内）
        const hasImagesInArticle = this.checkToutiaoImages(mobileResult.html);
        if (hasImagesInArticle) {
            if (this.verbose)
                console.error('[Fetch] 手机版有图片，使用手机版');
            return mobileResult;
        }
        // 3. 手机版无图片，尝试桌面版
        if (this.verbose)
            console.error('[Fetch] 手机版无图片，尝试桌面版...');
        const desktopResult = await this.fetchWithPlaywright(url, timeout, 'desktop');
        if (!desktopResult.success) {
            if (this.verbose)
                console.error('[Fetch] 桌面版抓取失败，返回手机版');
            return mobileResult;
        }
        // 4. 检查桌面版是否有图片
        const desktopHasImages = this.checkToutiaoImages(desktopResult.html);
        if (desktopHasImages) {
            if (this.verbose)
                console.error('[Fetch] 桌面版有图片，使用桌面版');
            return desktopResult;
        }
        // 5. 都没有图片，返回手机版
        if (this.verbose)
            console.error('[Fetch] 桌面版也无图片，返回手机版');
        return mobileResult;
    }
    /**
     * 检查头条页面 article 内是否有图片
     */
    checkToutiaoImages(html) {
        // 简单检查：article 标签内是否有 img 标签且 src 包含 toutiaoimg
        const articleMatch = html.match(/<article[^>]*>([\s\S]*?)<\/article>/i);
        if (!articleMatch)
            return false;
        const articleContent = articleMatch[1];
        // 检查是否有 toutiaoimg 图片
        return /<img[^>]+src=["'][^"']*toutiaoimg[^"']*["']/i.test(articleContent);
    }
    /**
     * 使用真实 Chrome 浏览器抓取（channel: 'chrome'）
     * 比 Playwright 自带的 Chromium 有更真实的浏览器指纹，可绕过知乎等平台的 zse-ck 反爬
     * 如果 Chrome 未安装，自动降级到普通 Playwright
     */
    async fetchWithRealChrome(url, timeout) {
        try {
            const playwright = await Promise.resolve().then(() => __importStar(require('playwright')));
            const browser = await playwright.chromium.launch({
                channel: 'chrome',
                headless: true,
                args: ['--disable-blink-features=AutomationControlled', '--no-sandbox'],
            });
            try {
                const context = await browser.newContext({
                    userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
                    locale: 'zh-CN',
                    viewport: { width: 1280, height: 800 },
                });
                await context.addInitScript(() => {
                    Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
                    Object.defineProperty(navigator, 'languages', { get: () => ['zh-CN', 'zh', 'en'] });
                });
                const page = await context.newPage();
                await page.goto(url, { waitUntil: 'domcontentloaded', timeout });
                await page.waitForTimeout(3000);
                const html = await page.content();
                if (html && html.length > 5000 && !html.includes('安全验证')) {
                    if (this.verbose)
                        console.error('[Fetch] ✅ 真实 Chrome 抓取成功');
                    return { html, method: 'playwright-chrome', success: true };
                }
                if (this.verbose)
                    console.error('[Fetch] 真实 Chrome 内容不足，降级到 Playwright Chromium');
            }
            finally {
                await browser.close();
            }
        }
        catch (e) {
            if (this.verbose)
                console.error('[Fetch] Chrome 不可用，降级到 Playwright Chromium:', e instanceof Error ? e.message : e);
        }
        return this.fetchWithPlaywright(url, timeout);
    }
    async fetchWithPlaywright(url, timeout, deviceType = 'desktop') {
        try {
            const playwright = await Promise.resolve().then(() => __importStar(require('playwright')));
            const browser = await playwright.chromium.launch({
                headless: true,
                args: [
                    '--disable-blink-features=AutomationControlled',
                    '--no-sandbox',
                    '--disable-setuid-sandbox',
                ]
            });
            try {
                // 根据 deviceType 选择 UA
                const isMobile = deviceType === 'mobile';
                const userAgent = isMobile
                    ? 'Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.0 Mobile/15E148 Safari/604.1'
                    : 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';
                const context = await browser.newContext({
                    viewport: isMobile ? { width: 375, height: 812 } : { width: 1920, height: 1080 },
                    userAgent,
                    deviceScaleFactor: isMobile ? 2 : 1,
                    isMobile,
                    hasTouch: isMobile,
                });
                // 注入更真实的浏览器指纹
                await context.addInitScript(() => {
                    // 绕过自动化检测
                    Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
                    // 添加一些常见的插件
                    Object.defineProperty(navigator, 'plugins', {
                        get: () => [1, 2, 3, 4, 5]
                    });
                    //  Languages
                    Object.defineProperty(navigator, 'languages', {
                        get: () => ['zh-CN', 'zh', 'en']
                    });
                });
                const page = await context.newPage();
                // 等待页面加载完成
                await page.goto(url, {
                    waitUntil: 'networkidle',
                    timeout,
                });
                // 根据不同平台等待不同的选择器
                const isToutiao = url.includes('toutiao.com');
                const isWechat = url.includes('mp.weixin.qq.com');
                try {
                    if (isToutiao) {
                        // 头条页面：等待文章内容加载
                        await page.waitForSelector('article, .article-content, .syl-article-base', { timeout: 8000 });
                        // 滚动页面以触发懒加载图片
                        await page.evaluate(() => {
                            window.scrollTo(0, document.body.scrollHeight / 2);
                        });
                        await page.waitForTimeout(1000);
                        await page.evaluate(() => {
                            window.scrollTo(0, document.body.scrollHeight);
                        });
                        await page.waitForTimeout(1000);
                        // 等待图片加载
                        await page.waitForSelector('article img', { timeout: 5000 }).catch(() => { });
                    }
                    else if (isWechat) {
                        await page.waitForSelector('#js_content, .weui-msg', { timeout: 5000 });
                    }
                    else {
                        await page.waitForSelector('article, .article, .content, main', { timeout: 5000 });
                    }
                }
                catch (e) {
                    // 忽略选择器超时
                }
                // 等待额外内容加载
                await page.waitForTimeout(2000);
                const html = await page.content();
                // 检测是否触发了验证页面
                if (html.includes('环境异常') || html.includes('js_verify')) {
                    return {
                        html: '',
                        method: 'playwright',
                        success: false,
                        error: '微信验证页面，需要登录或更换网络环境'
                    };
                }
                return { html, method: 'playwright', success: true };
            }
            finally {
                await browser.close();
            }
        }
        catch (error) {
            return {
                html: '',
                method: 'playwright',
                success: false,
                error: error instanceof Error ? error.message : String(error)
            };
        }
    }
}
exports.Fetcher = Fetcher;
