// news-to-markdown/dist/utils/image-downloader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadImage = downloadImage;
exports.downloadImages = downloadImages;
exports.replaceImageUrls = replaceImageUrls;
const promises_1 = require("fs/promises");
const path_1 = require("path");
const crypto_1 = __importDefault(require("crypto"));
const https_1 = __importDefault(require("https"));
const http_1 = __importDefault(require("http"));
/**
 * 使用原生 http/https 下载文件
 */
function downloadWithHttp(url, timeout) {
    return new Promise((resolve, reject) => {
        const urlObj = new URL(url);
        const client = urlObj.protocol === 'https:' ? https_1.default : http_1.default;
        const request = client.get(url, {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
            },
            timeout,
        }, (response) => {
            if (response.statusCode === 301 || response.statusCode === 302) {
                const redirectUrl = response.headers.location;
                if (redirectUrl) {
                    downloadWithHttp(redirectUrl, timeout).then(resolve).catch(reject);
                    return;
                }
            }
            if (response.statusCode !== 200) {
                reject(new Error(`HTTP ${response.statusCode}`));
                return;
            }
            const chunks = [];
            response.on('data', (chunk) => chunks.push(chunk));
            response.on('end', () => resolve(Buffer.concat(chunks)));
            response.on('error', reject);
        });
        request.on('error', reject);
        request.on('timeout', () => {
            request.destroy();
            reject(new Error('下载超时'));
        });
    });
}
/**
 * 下载图片到本地
 */
async function downloadImage(options) {
    const { url, outputDir, filename, timeout = 30000, maxRetries = 3 } = options;
    try {
        await (0, promises_1.mkdir)(outputDir, { recursive: true });
    }
    catch (error) {
        return {
            success: false,
            error: `创建目录失败: ${error}`,
        };
    }
    const finalFilename = filename || generateFilename(url);
    const localPath = (0, path_1.join)(outputDir, finalFilename);
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
        try {
            const buffer = await downloadWithHttp(url, timeout);
            if (!buffer || buffer.length === 0) {
                throw new Error('下载的图片大小为 0');
            }
            await (0, promises_1.writeFile)(localPath, buffer);
            return {
                success: true,
                localPath,
            };
        }
        catch (error) {
            if (attempt === maxRetries) {
                return {
                    success: false,
                    error: `下载失败 (${maxRetries} 次重试): ${error.message}`,
                };
            }
            await sleep(1000 * attempt);
        }
    }
    return {
        success: false,
        error: '下载失败',
    };
}
/**
 * 批量下载图片
 */
async function downloadImages(urls, outputDir, verbose = false) {
    const urlToLocalPath = new Map();
    await (0, promises_1.mkdir)(outputDir, { recursive: true });
    const downloadPromises = urls.map(async (url, index) => {
        if (verbose)
            console.error(`[ImageDownloader] 下载图片 ${index + 1}/${urls.length}: ${url}`);
        const result = await downloadImage({
            url,
            outputDir,
            filename: `image_${index + 1}${getExtensionFromUrl(url)}`,
        });
        if (result.success && result.localPath) {
            urlToLocalPath.set(url, result.localPath);
            if (verbose)
                console.error(`[ImageDownloader] ✓ 下载成功: ${result.localPath}`);
        }
        else {
            if (verbose)
                console.error(`[ImageDownloader] ✗ 下载失败: ${result.error}`);
        }
    });
    await Promise.all(downloadPromises);
    return urlToLocalPath;
}
/**
 * 从 URL 生成文件名
 */
function generateFilename(url) {
    const urlWithoutQuery = url.split('?')[0];
    const urlPath = new URL(url).pathname;
    const basename = urlPath.split('/').pop() || 'image';
    const ext = (0, path_1.extname)(basename);
    if (ext && ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp'].includes(ext.toLowerCase())) {
        return basename;
    }
    const hash = crypto_1.default.createHash('md5').update(url).digest('hex').substring(0, 12);
    return `${hash}.jpg`;
}
/**
 * 从 URL 获取扩展名
 */
function getExtensionFromUrl(url) {
    const urlWithoutQuery = url.split('?')[0];
    const ext = (0, path_1.extname)(urlWithoutQuery).toLowerCase();
    if (['.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp'].includes(ext)) {
        return ext;
    }
    return '.jpg';
}
/**
 * 延迟函数
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
/**
 * 替换 Markdown 中的图片 URL 为本地路径
 */
function replaceImageUrls(markdown, urlToLocalPath, baseDir) {
    let result = markdown;
    for (const [url, localPath] of urlToLocalPath.entries()) {
        // 使用 images/ 目录前缀
        const filename = localPath.split('/').pop();
        const relativePath = `./images/${filename}`;
        // 提取 URL 的基础部分（去掉 #imgIndex 和查询参数）
        const baseUrl = url.split('#')[0].split('?')[0];
        // 匹配完整 URL（包含可能的查询参数和 #imgIndex）
        const imageRegex = new RegExp(`!\\[([^\\]]*)\\]\\(${escapeRegex(url)}\\)`, 'g');
        result = result.replace(imageRegex, `![$1](${relativePath})`);
        // 也匹配不带 #imgIndex 的 URL
        const urlWithoutHash = url.split('#')[0];
        if (urlWithoutHash !== url) {
            const imageRegex2 = new RegExp(`!\\[([^\\]]*)\\]\\(${escapeRegex(urlWithoutHash)}\\)`, 'g');
            result = result.replace(imageRegex2, `![$1](${relativePath})`);
        }
        // 匹配基础 URL（不带查询参数）
        if (baseUrl !== urlWithoutHash) {
            const imageRegex3 = new RegExp(`!\\[([^\\]]*)\\]\\(${escapeRegex(baseUrl)}[^)]*\\)`, 'g');
            result = result.replace(imageRegex3, `![$1](${relativePath})`);
        }
    }
    return result;
}
/**
 * 转义正则表达式特殊字符
 */
function escapeRegex(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
