// @siping/html-to-markdown-node/dist/plugins/commonmark.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonMarkPlugin = void 0;
const Register_1 = require("../Register");
const dom_1 = require("../utils/dom");
const url_1 = require("../utils/url");
class CommonMarkPlugin {
    constructor(options = {}) {
        this.name = 'commonmark';
        this.options = {
            strongDelimiter: options.strongDelimiter || '**',
            emDelimiter: options.emDelimiter || '*',
            codeDelimiter: options.codeDelimiter || '`',
            headingStyle: options.headingStyle || 'atx'
        };
    }
    init(converter) {
        converter.register.escapedChar('*', '_', '[', ']', '(', ')', '#', '+', '-', '.', '!', '`', '\\', '|', '~', '>', '<');
        converter.register.renderer(this.renderStrong.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderEm.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderCode.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderPre.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderHeading.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderParagraph.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderBr.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderHr.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderLink.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderImage.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderBlockquote.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderList.bind(this), Register_1.PriorityStandard);
        converter.register.renderer(this.renderListItem.bind(this), Register_1.PriorityStandard);
    }
    renderStrong(ctx, node) {
        const tagName = (0, dom_1.getNodeName)(node);
        if (tagName !== 'strong' && tagName !== 'b') {
            return 'try-next';
        }
        const content = ctx.convertNode(node);
        if (!content)
            return '';
        const delim = this.options.strongDelimiter;
        return `${delim}${content}${delim}`;
    }
    renderEm(ctx, node) {
        const tagName = (0, dom_1.getNodeName)(node);
        if (tagName !== 'em' && tagName !== 'i') {
            return 'try-next';
        }
        const content = ctx.convertNode(node);
        if (!content)
            return '';
        const delim = this.options.emDelimiter;
        return `${delim}${content}${delim}`;
    }
    renderCode(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'code') {
            return 'try-next';
        }
        if (node.parentNode && (0, dom_1.getNodeName)(node.parentNode) === 'pre') {
            return 'try-next';
        }
        const content = (0, dom_1.getTextContent)(node);
        if (!content)
            return '';
        const delim = this.options.codeDelimiter;
        const backticks = content.includes('`') ? '``' : '`';
        const space = content.startsWith('`') || content.endsWith('`') ? ' ' : '';
        return `${backticks}${space}${content}${space}${backticks}`;
    }
    renderPre(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'pre') {
            return 'try-next';
        }
        let content = (0, dom_1.getTextContent)(node);
        if (!content)
            return '';
        const fence = '```';
        return `\n${fence}\n${content}\n${fence}\n`;
    }
    renderHeading(ctx, node) {
        const tagName = (0, dom_1.getNodeName)(node);
        const match = tagName.match(/^h([1-6])$/);
        if (!match) {
            return 'try-next';
        }
        const level = parseInt(match[1]);
        const content = ctx.convertNode(node);
        if (!content)
            return '';
        if (this.options.headingStyle === 'atx' || level > 2) {
            const hashes = '#'.repeat(level);
            return `\n${hashes} ${content}\n`;
        }
        else {
            const underline = level === 1 ? '=' : '-';
            return `\n${content}\n${underline.repeat(content.length)}\n`;
        }
    }
    renderParagraph(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'p') {
            return 'try-next';
        }
        const content = ctx.convertNode(node);
        if (!content)
            return '';
        return `\n${content}\n`;
    }
    renderBr(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'br') {
            return 'try-next';
        }
        return '  \n';
    }
    renderHr(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'hr') {
            return 'try-next';
        }
        return '\n---\n';
    }
    renderLink(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'a') {
            return 'try-next';
        }
        const href = (0, dom_1.getAttribute)(node, 'href') || '';
        const title = (0, dom_1.getAttribute)(node, 'title');
        const content = ctx.convertNode(node);
        const url = (0, url_1.resolveURL)(href, ctx.domain);
        const escapedUrl = (0, url_1.escapeMarkdownURL)(url);
        if (title) {
            return `[${content}](${escapedUrl} "${title}")`;
        }
        return `[${content}](${escapedUrl})`;
    }
    renderImage(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'img') {
            return 'try-next';
        }
        const src = (0, dom_1.getAttribute)(node, 'src') || '';
        const alt = (0, dom_1.getAttribute)(node, 'alt') || '';
        const title = (0, dom_1.getAttribute)(node, 'title');
        const url = (0, url_1.resolveURL)(src, ctx.domain);
        const escapedUrl = (0, url_1.escapeMarkdownURL)(url);
        if (title) {
            return `![${alt}](${escapedUrl} "${title}")`;
        }
        return `![${alt}](${escapedUrl})`;
    }
    renderBlockquote(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'blockquote') {
            return 'try-next';
        }
        const content = ctx.convertNode(node);
        if (!content)
            return '';
        const lines = content.split('\n').filter((line) => line.trim());
        const quoted = lines.map((line) => `> ${line}`).join('\n');
        return `\n${quoted}\n`;
    }
    renderList(ctx, node) {
        const tagName = (0, dom_1.getNodeName)(node);
        if (tagName !== 'ul' && tagName !== 'ol') {
            return 'try-next';
        }
        const content = ctx.convertNode(node);
        if (!content)
            return '';
        return `\n${content}\n`;
    }
    renderListItem(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'li') {
            return 'try-next';
        }
        const content = ctx.convertNode(node);
        if (!content)
            return '';
        const parent = node.parentNode;
        const isOrdered = parent && (0, dom_1.getNodeName)(parent) === 'ol';
        if (isOrdered) {
            return `1. ${content.trim()}\n`;
        }
        else {
            return `- ${content.trim()}\n`;
        }
    }
}
exports.CommonMarkPlugin = CommonMarkPlugin;
