// @siping/html-to-markdown-node/dist/plugins/table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TablePlugin = void 0;
const Register_1 = require("../Register");
const dom_1 = require("../utils/dom");
class TablePlugin {
    constructor() {
        this.name = 'table';
    }
    init(converter) {
        converter.register.renderer(this.renderTable.bind(this), Register_1.PriorityStandard);
    }
    renderTable(ctx, node) {
        if ((0, dom_1.getNodeName)(node) !== 'table') {
            return 'try-next';
        }
        const rows = [];
        let hasHeader = false;
        // Extract table data
        const thead = this.findChild(node, 'thead');
        const tbody = this.findChild(node, 'tbody');
        const tfoot = this.findChild(node, 'tfoot');
        // Process header
        if (thead && (0, dom_1.hasChildNodes)(thead)) {
            const headerRows = this.extractRows(ctx, thead);
            if (headerRows.length > 0) {
                rows.push(...headerRows);
                hasHeader = true;
            }
        }
        // Process body
        if (tbody && (0, dom_1.hasChildNodes)(tbody)) {
            const bodyRows = this.extractRows(ctx, tbody);
            rows.push(...bodyRows);
        }
        else {
            // If no tbody, process direct tr children
            const directRows = this.extractRows(ctx, node);
            if (!hasHeader && directRows.length > 0) {
                // First row might be header if it contains th elements
                const firstRow = node.querySelector('tr');
                if (firstRow && firstRow.querySelector('th')) {
                    hasHeader = true;
                }
            }
            rows.push(...directRows);
        }
        // Process footer
        if (tfoot && (0, dom_1.hasChildNodes)(tfoot)) {
            const footerRows = this.extractRows(ctx, tfoot);
            rows.push(...footerRows);
        }
        if (rows.length === 0) {
            return '';
        }
        // Build markdown table
        const markdown = [];
        // Add header row
        const headerRow = rows[0];
        markdown.push('| ' + headerRow.join(' | ') + ' |');
        // Add separator
        const separator = headerRow.map(() => '---').join(' | ');
        markdown.push('| ' + separator + ' |');
        // Add data rows
        for (let i = 1; i < rows.length; i++) {
            const row = rows[i];
            // Ensure row has same number of columns as header
            while (row.length < headerRow.length) {
                row.push('');
            }
            markdown.push('| ' + row.join(' | ') + ' |');
        }
        return '\n' + markdown.join('\n') + '\n';
    }
    findChild(node, tagName) {
        if (!node.children)
            return null;
        for (let i = 0; i < node.children.length; i++) {
            const child = node.children[i];
            if ((0, dom_1.getNodeName)(child) === tagName) {
                return child;
            }
        }
        return null;
    }
    extractRows(ctx, parent) {
        const rows = [];
        if (!parent.children)
            return rows;
        for (let i = 0; i < parent.children.length; i++) {
            const child = parent.children[i];
            if ((0, dom_1.getNodeName)(child) === 'tr') {
                const cells = this.extractCells(ctx, child);
                if (cells.length > 0) {
                    rows.push(cells);
                }
            }
        }
        return rows;
    }
    extractCells(ctx, tr) {
        const cells = [];
        if (!tr.children)
            return cells;
        for (let i = 0; i < tr.children.length; i++) {
            const child = tr.children[i];
            const tagName = (0, dom_1.getNodeName)(child);
            if (tagName === 'td' || tagName === 'th') {
                // Convert cell content to markdown
                const content = ctx.convertNode(child);
                // Clean up the content: remove newlines and extra spaces
                const cleaned = content.replace(/\n/g, ' ').replace(/\s+/g, ' ').trim();
                cells.push(cleaned || ' ');
            }
        }
        return cells;
    }
}
exports.TablePlugin = TablePlugin;
