// news-to-markdown/dist/cli.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const converter_1 = require("./core/converter");
const fs_1 = require("fs");
const path_1 = require("path");
function parseArgs(args) {
    const options = { url: '' };
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === '-o' || arg === '--output') {
            options.output = args[++i];
        }
        else if (arg === '-p' || arg === '--platform') {
            options.platform = args[++i];
        }
        else if (arg === '-d' || arg === '--download-images') {
            options.downloadImages = true;
        }
        else if (arg === '-v' || arg === '--verbose') {
            options.verbose = true;
        }
        else if (arg === '--json') {
            options.json = true;
        }
        else if (arg === '-h' || arg === '--help') {
            printHelp();
            process.exit(0);
        }
        else if (arg === '--version') {
            console.log(require('../package.json').version);
            process.exit(0);
        }
        else if (!arg.startsWith('-') && !options.url) {
            options.url = arg;
        }
    }
    return options;
}
function printHelp() {
    console.log(`
news-to-markdown - Convert news articles to Markdown

Usage:
  news-to-markdown <url> [options]

Options:
  -o, --output <file>     Output file path (default: stdout)
  -p, --platform <name>   Force platform (auto-detect if not specified)
  -d, --download-images   Download images to local directory
  -v, --verbose           Show verbose output
  --json                  Output metadata as JSON
  -h, --help              Show this help
  --version               Show version

Examples:
  news-to-markdown https://mp.weixin.qq.com/s/xxx
  news-to-markdown https://toutiao.com/xxx -o article.md
  news-to-markdown https://xiaohongshu.com/xxx -d -o ./output/article.md
`);
}
async function main() {
    const args = process.argv.slice(2);
    const options = parseArgs(args);
    if (!options.url) {
        console.error('Error: URL is required');
        printHelp();
        process.exit(1);
    }
    const converter = new converter_1.NewsToMarkdownConverter();
    try {
        const result = await converter.convert({
            url: options.url,
            platform: options.platform,
            downloadImages: options.downloadImages,
            outputDir: options.output ? (0, path_1.dirname)(options.output) : undefined,
            verbose: options.verbose,
        });
        if (options.json) {
            console.log(JSON.stringify(result.metadata, null, 2));
        }
        else if (options.output) {
            const outputDir = (0, path_1.dirname)(options.output);
            (0, fs_1.mkdirSync)(outputDir, { recursive: true });
            (0, fs_1.writeFileSync)(options.output, result.markdown);
            if (options.verbose) {
                console.error(`Written to: ${options.output}`);
            }
        }
        else {
            console.log(result.markdown);
        }
    }
    catch (error) {
        console.error('Error:', error instanceof Error ? error.message : error);
        process.exit(1);
    }
}
main();
