// news-to-markdown/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformRegistry = exports.XiaoHongShuPlatform = exports.WechatPlatform = exports.TouTiaoPlatform = exports.GenericPlatform = exports.BasePlatform = exports.DualExtractor = exports.Fetcher = exports.NewsToMarkdownConverter = void 0;
var converter_1 = require("./core/converter");
Object.defineProperty(exports, "NewsToMarkdownConverter", { enumerable: true, get: function () { return converter_1.NewsToMarkdownConverter; } });
var fetcher_1 = require("./core/fetcher");
Object.defineProperty(exports, "Fetcher", { enumerable: true, get: function () { return fetcher_1.Fetcher; } });
var dual_extractor_1 = require("./core/dual-extractor");
Object.defineProperty(exports, "DualExtractor", { enumerable: true, get: function () { return dual_extractor_1.DualExtractor; } });
var platforms_1 = require("./platforms");
Object.defineProperty(exports, "BasePlatform", { enumerable: true, get: function () { return platforms_1.BasePlatform; } });
Object.defineProperty(exports, "GenericPlatform", { enumerable: true, get: function () { return platforms_1.GenericPlatform; } });
Object.defineProperty(exports, "TouTiaoPlatform", { enumerable: true, get: function () { return platforms_1.TouTiaoPlatform; } });
Object.defineProperty(exports, "WechatPlatform", { enumerable: true, get: function () { return platforms_1.WechatPlatform; } });
Object.defineProperty(exports, "XiaoHongShuPlatform", { enumerable: true, get: function () { return platforms_1.XiaoHongShuPlatform; } });
Object.defineProperty(exports, "PlatformRegistry", { enumerable: true, get: function () { return platforms_1.PlatformRegistry; } });
__exportStar(require("./types"), exports);
