// news-to-markdown/dist/platforms/36kr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Kr36Platform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
/**
 * 36kr 平台适配器
 *
 * 注意：36kr 电脑端页面有反爬机制，需要使用移动端页面
 * 移动端 URL 格式：https://m.36kr.com/p/{article_id}
 * 电脑端 URL 格式：https://36kr.com/p/{article_id}
 */
class Kr36Platform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = '36kr';
        this.metaSelectors = {
            title: 'meta[property="og:title"], .article-title, .title, h1',
            author: 'meta[name="author"], .author-name, .article-author, .user-name, .author',
            time: 'meta[property="article:published_time"], time[datetime], .publish-time, .time, .date',
        };
    }
    // 36kr 特有的噪音选择器
    get noiseSelectors() {
        return [
            ...this.commonNoiseSelectors,
            '.fixed-bar', '.tool-bar',
            '.article-bottom', '.article-footer',
            '.kr-flow-bar', '.kr-side-bar',
            '.article-recommend', '.recommend-list',
            '.download-app', '.app-download',
            '.article-tags', '.tags',
            '.article-action', '.action-bar',
        ];
    }
    // 36kr 内容选择器（优先级从高到低）
    get contentSelectors() {
        return [
            '.article-content',
            '.article-detail-content',
            '.articleDetailContent',
            '.article-body',
            '.content',
            '.newsflash-content',
            '.newsflash-item-content',
            'article',
            '.post-content',
            '.common-width',
        ];
    }
    detect(url) {
        return this.matchDomain(url, ['36kr.com']);
    }
    /**
     * 将电脑端 URL 转换为移动端 URL
     * 36kr 电脑端有反爬机制，移动端更容易获取内容
     */
    getMobileUrl(url) {
        try {
            const urlObj = new URL(url);
            if (urlObj.hostname.startsWith('m.')) {
                return url;
            }
            urlObj.hostname = 'm.' + urlObj.hostname.replace(/^www\./, '');
            return urlObj.toString();
        }
        catch {
            return url;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // 移除噪音元素
        this.removeNoiseElements($, this.noiseSelectors);
        // 查找内容
        const $content = this.findContent($, this.contentSelectors);
        if ($content) {
            this.processLazyImages($, $content);
            this.removeEmptyCaptions($, $content);
            this.cleanEmptyTags($, $content);
            return $.html($content);
        }
        // Fallback
        const $body = $('body');
        this.processLazyImages($, $body);
        this.cleanEmptyTags($, $body);
        return $body.html() || html;
    }
    postprocess(markdown) {
        return this.cleanEscapes(markdown);
    }
}
exports.Kr36Platform = Kr36Platform;
