// news-to-markdown/dist/platforms/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasePlatform = void 0;
class BasePlatform {
    preprocess(html, url) {
        return html;
    }
    postprocess(markdown) {
        return markdown;
    }
    /**
     * 通用方法：检测 URL 是否匹配指定域名
     * @param url 要检测的 URL
     * @param domains 域名列表，如 ['36kr.com', 'm.36kr.com']
     */
    matchDomain(url, domains) {
        try {
            const hostname = new URL(url).hostname.replace(/^www\./, '');
            return domains.some(d => hostname.includes(d));
        }
        catch {
            return false;
        }
    }
    /**
     * 通用方法：移除噪音元素
     * @param $ Cheerio 实例
     * @param selectors 要移除的选择器列表
     */
    removeNoiseElements($, selectors) {
        for (const sel of selectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
    }
    /**
     * 通用噪音选择器（可被子类扩展）
     */
    get commonNoiseSelectors() {
        return [
            'header', 'footer', 'nav', 'aside',
            '.nav', '.navbar', '.header', '.footer',
            '.sidebar', '.side-bar', '#sidebar',
            '.recommend', '.related', '.comment', '.comments',
            '.share', '.share-bar', '.social-share',
            '.ad', '.advertisement', '[class*="ad-"]',
            'script', 'style', 'noscript', 'iframe',
        ];
    }
    /**
     * 通用方法：处理懒加载图片
     * 将 data-src、data-original 等属性转换为 src
     * @param $ Cheerio 实例
     * @param $root 根元素
     * @param dataAttrs 懒加载属性名列表
     */
    processLazyImages($, $root, dataAttrs) {
        const attrs = dataAttrs || ['data-src', 'data-original', 'data-lazy-src', 'data-actualsrc', 'data-original-src'];
        $root.find('img').each((_i, el) => {
            const $img = $(el);
            // 尝试从各种 data 属性获取真实图片 URL
            for (const attr of attrs) {
                const dataSrc = $img.attr(attr);
                if (dataSrc && !dataSrc.startsWith('data:image')) {
                    $img.attr('src', dataSrc);
                    break;
                }
            }
            // 移除占位图片
            const src = $img.attr('src') || '';
            if (src.startsWith('data:image') || !src) {
                $img.remove();
            }
        });
    }
    /**
     * 通用方法：移除空的图片说明
     * @param $ Cheerio 实例
     * @param $root 根元素
     * @param captionSelectors 图片说明选择器列表
     */
    removeEmptyCaptions($, $root, captionSelectors) {
        const selectors = captionSelectors || ['figcaption', '.image-caption', '.img-caption', '.caption'];
        for (const sel of selectors) {
            $root.find(sel).each((_i, el) => {
                const $caption = $(el);
                if ($caption.text().trim().length === 0) {
                    $caption.remove();
                }
            });
        }
    }
    /**
     * 通用方法：尝试从多个选择器中提取内容
     * @param $ Cheerio 实例
     * @param selectors 内容选择器列表（按优先级排序）
     * @param minLength 最小内容长度
     * @returns 找到的内容元素，或 null
     */
    findContent($, selectors, minLength = 50) {
        for (const sel of selectors) {
            const $content = $(sel).first();
            if ($content.length && $content.text().trim().length >= minLength) {
                return $content;
            }
        }
        return null;
    }
    /**
     * 通用方法：清理空标签（包括带 visibility 样式的占位符）
     * 适用于所有平台，防止空 <p>、<div>、<span> 等标签污染输出
     * 注意：保留只含图片的元素
     */
    cleanEmptyTags($, $root) {
        $root.find('p, div, span, section').each((i, el) => {
            const $el = $(el);
            const text = $el.text().trim();
            const hasChildren = $el.children().length > 0;
            const hasImg = $el.find('img').length > 0;
            // 如果元素为空（无文字且无子元素），移除
            if (!text && !hasChildren) {
                $el.remove();
                return;
            }
            // 如果元素只有图片，保留
            if (hasImg && !text) {
                return;
            }
            // 特别处理：如果元素只有 visibility 或 display 样式且为空，也移除
            const style = $el.attr('style') || '';
            if ((style.includes('visibility') || style.includes('display: none')) && !text && !hasChildren) {
                $el.remove();
            }
        });
    }
    /**
     * 通用方法：清理 Markdown 中不必要的转义字符
     */
    cleanEscapes(markdown) {
        return markdown
            .replace(/\\-/g, '-')
            .replace(/\\\./g, '.')
            .replace(/\n{4,}/g, '\n\n\n');
    }
}
exports.BasePlatform = BasePlatform;
