// news-to-markdown/dist/platforms/bilibili.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BilibiliPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class BilibiliPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'bilibili';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.title, h1',
            author: 'meta[name="author"], .author-name, .up-name',
            time: 'meta[property="article:published_time"], time[datetime], .publish-time',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('bilibili.com') && (url.includes('/opus/') || url.includes('/read/'));
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.sidebar', '.recommend', '.comment',
            '.related', '.share-panel',
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('.opus-detail, .article-content, article').first();
        if ($article.length) {
            this.processBilibiliContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    postprocess(markdown) {
        // Clean title suffix
        markdown = markdown.replace(/\s*[–-]\s*哔哩哔哩\s*/g, '');
        // Clean author field - remove date suffix
        markdown = markdown.replace(/^(\*\*作者\*\*:\s+.+?)\s+\d{4}年\d{2}月\d{2}日\s+\d{2}:\d{2}/gm, '$1');
        return markdown;
    }
    processBilibiliContent($, $content) {
        // Process images - fix protocol-relative URLs
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            let src = $img.attr('src') || $img.attr('data-src');
            // Fix protocol-relative URLs
            if (src && src.startsWith('//')) {
                src = 'https:' + src;
                $img.attr('src', src);
            }
        });
    }
}
exports.BilibiliPlatform = BilibiliPlatform;
