// news-to-markdown/dist/platforms/cnblogs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CnblogsPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class CnblogsPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'cnblogs';
        this.metaSelectors = {
            title: 'meta[property="og:title"], #cb_post_title_url, h1.postTitle, h1',
            author: 'meta[name="author"], #author_profile_info a, .postDesc a',
            time: 'meta[property="article:published_time"], #post-date, .postDesc',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('cnblogs.com');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.sidebar', '.recommend', '.comment',
            '.ad', '.advertisement',
            '#blog-sidecolumn', // 博客园侧边栏
            '#blog-comments-placeholder', // 评论区
            '.c_ad_block', // 广告
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('#cnblogs_post_body, .post-body, article').first();
        if ($article.length) {
            this.processCnblogsContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    processCnblogsContent($, $content) {
        // Process images
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-src') || $img.attr('data-original');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
        // Remove code copy buttons
        $content.find('.cnblogs_code_copy, .code-copy').remove();
    }
}
exports.CnblogsPlatform = CnblogsPlatform;
