// news-to-markdown/dist/platforms/csdn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CSDNPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class CSDNPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'csdn';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.title-article, h1',
            author: 'meta[name="author"], .follow-nickName, .user-info .username, .author-name',
            time: 'meta[property="article:published_time"], .article-bar-top .time, time[datetime]',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('csdn.net');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.side-toolbar', '.recommend', '.comment',
            '.tool-box', '.article-bar-bottom',
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('#content_views, #article_content, .markdown_views, article').first();
        if ($article.length) {
            this.processCSDNContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    processCSDNContent($, $content) {
        // Process images
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-original-src') || $img.attr('data-src');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
        // Remove code copy buttons
        $content.find('.hljs-button, .copy-code-btn').remove();
        // Remove "一键三连" and other promotional elements
        $content.find('.recommend-box, .article-bar').remove();
    }
}
exports.CSDNPlatform = CSDNPlatform;
