// news-to-markdown/dist/platforms/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformRegistry = exports.Kr36Platform = exports.CnblogsPlatform = exports.SegmentFaultPlatform = exports.BilibiliPlatform = exports.OschinaPlatform = exports.WoshipmPlatform = exports.CSDNPlatform = exports.JianshuPlatform = exports.JuejinPlatform = exports.ZhihuPlatform = exports.XiaoHongShuPlatform = exports.WechatPlatform = exports.TouTiaoPlatform = exports.GenericPlatform = exports.BasePlatform = void 0;
var base_1 = require("./base");
Object.defineProperty(exports, "BasePlatform", { enumerable: true, get: function () { return base_1.BasePlatform; } });
var generic_1 = require("./generic");
Object.defineProperty(exports, "GenericPlatform", { enumerable: true, get: function () { return generic_1.GenericPlatform; } });
var toutiao_1 = require("./toutiao");
Object.defineProperty(exports, "TouTiaoPlatform", { enumerable: true, get: function () { return toutiao_1.TouTiaoPlatform; } });
var wechat_1 = require("./wechat");
Object.defineProperty(exports, "WechatPlatform", { enumerable: true, get: function () { return wechat_1.WechatPlatform; } });
var xiaohongshu_1 = require("./xiaohongshu");
Object.defineProperty(exports, "XiaoHongShuPlatform", { enumerable: true, get: function () { return xiaohongshu_1.XiaoHongShuPlatform; } });
var zhihu_1 = require("./zhihu");
Object.defineProperty(exports, "ZhihuPlatform", { enumerable: true, get: function () { return zhihu_1.ZhihuPlatform; } });
var juejin_1 = require("./juejin");
Object.defineProperty(exports, "JuejinPlatform", { enumerable: true, get: function () { return juejin_1.JuejinPlatform; } });
var jianshu_1 = require("./jianshu");
Object.defineProperty(exports, "JianshuPlatform", { enumerable: true, get: function () { return jianshu_1.JianshuPlatform; } });
var csdn_1 = require("./csdn");
Object.defineProperty(exports, "CSDNPlatform", { enumerable: true, get: function () { return csdn_1.CSDNPlatform; } });
var woshipm_1 = require("./woshipm");
Object.defineProperty(exports, "WoshipmPlatform", { enumerable: true, get: function () { return woshipm_1.WoshipmPlatform; } });
var oschina_1 = require("./oschina");
Object.defineProperty(exports, "OschinaPlatform", { enumerable: true, get: function () { return oschina_1.OschinaPlatform; } });
var bilibili_1 = require("./bilibili");
Object.defineProperty(exports, "BilibiliPlatform", { enumerable: true, get: function () { return bilibili_1.BilibiliPlatform; } });
var segmentfault_1 = require("./segmentfault");
Object.defineProperty(exports, "SegmentFaultPlatform", { enumerable: true, get: function () { return segmentfault_1.SegmentFaultPlatform; } });
var cnblogs_1 = require("./cnblogs");
Object.defineProperty(exports, "CnblogsPlatform", { enumerable: true, get: function () { return cnblogs_1.CnblogsPlatform; } });
var _36kr_1 = require("./36kr");
Object.defineProperty(exports, "Kr36Platform", { enumerable: true, get: function () { return _36kr_1.Kr36Platform; } });
var registry_1 = require("./registry");
Object.defineProperty(exports, "PlatformRegistry", { enumerable: true, get: function () { return registry_1.PlatformRegistry; } });
