// news-to-markdown/dist/platforms/jianshu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JianshuPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class JianshuPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'jianshu';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.title, h1',
            author: 'meta[name="author"], .author .name, .author-name, a.author',
            time: 'meta[property="article:published_time"], time[datetime]',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('jianshu.com');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.side-tool', '.recommend', '.related',
            '.comment', '.comments',
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('article, .show-content, .article').first();
        if ($article.length) {
            this.processJianshuContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    processJianshuContent($, $content) {
        // Process images
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-original-src') || $img.attr('data-src');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
        // Remove image captions if empty
        $content.find('.image-caption').each((_i, el) => {
            const $caption = $(el);
            if ($caption.text().trim().length === 0) {
                $caption.remove();
            }
        });
    }
}
exports.JianshuPlatform = JianshuPlatform;
