// news-to-markdown/dist/platforms/juejin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JuejinPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class JuejinPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'juejin';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.article-title, h1',
            author: 'meta[name="author"], .username, .author-name',
            time: 'meta[property="article:published_time"], time[datetime]',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('juejin.cn');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.sidebar', '.side-bar', '.comment',
            '.recommend', '.related',
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('.markdown-body, article, .article-content').first();
        if ($article.length) {
            this.processJuejinContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    processJuejinContent($, $content) {
        // Process images - Juejin uses lazy loading
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-src') || $img.attr('data-original');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
        // Remove code copy buttons
        $content.find('.copy-code-btn, .code-block-extension-codeCopyBtn').remove();
    }
}
exports.JuejinPlatform = JuejinPlatform;
