// news-to-markdown/dist/platforms/oschina.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OschinaPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class OschinaPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'oschina';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.article-title, h1',
            author: 'meta[name="author"], .author-name, .author',
            time: 'meta[property="article:published_time"], time[datetime], .pub-date',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('oschina.net');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.sidebar', '.recommend', '.comment',
            '.ad', '.advertisement',
            '.news-content-info', // 移除顶部信息栏
            '.news-content-AiDesc-title', // 移除 AI 辅阅标题
            '.extra-feature', // 移除额外功能
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content - 开源中国使用 .news-content
        const $article = $('.news-content, .article-detail, .news-body, article').first();
        if ($article.length) {
            this.processOschinaContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    processOschinaContent($, $content) {
        // Process images
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-src') || $img.attr('data-original');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
        // Remove code copy buttons
        $content.find('.copy-btn, .code-copy').remove();
    }
}
exports.OschinaPlatform = OschinaPlatform;
