// news-to-markdown/dist/platforms/registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformRegistry = void 0;
const generic_1 = require("./generic");
const toutiao_1 = require("./toutiao");
const wechat_1 = require("./wechat");
const xiaohongshu_1 = require("./xiaohongshu");
const zhihu_1 = require("./zhihu");
const juejin_1 = require("./juejin");
const jianshu_1 = require("./jianshu");
const csdn_1 = require("./csdn");
const woshipm_1 = require("./woshipm");
const oschina_1 = require("./oschina");
const bilibili_1 = require("./bilibili");
const segmentfault_1 = require("./segmentfault");
const cnblogs_1 = require("./cnblogs");
const _36kr_1 = require("./36kr");
class PlatformRegistry {
    constructor() {
        this.platforms = new Map();
        this.registerBuiltinPlatforms();
    }
    registerBuiltinPlatforms() {
        this.register(new toutiao_1.TouTiaoPlatform());
        this.register(new wechat_1.WechatPlatform());
        this.register(new xiaohongshu_1.XiaoHongShuPlatform());
        this.register(new zhihu_1.ZhihuPlatform());
        this.register(new juejin_1.JuejinPlatform());
        this.register(new jianshu_1.JianshuPlatform());
        this.register(new csdn_1.CSDNPlatform());
        this.register(new woshipm_1.WoshipmPlatform());
        this.register(new oschina_1.OschinaPlatform());
        this.register(new bilibili_1.BilibiliPlatform());
        this.register(new segmentfault_1.SegmentFaultPlatform());
        this.register(new cnblogs_1.CnblogsPlatform());
        this.register(new _36kr_1.Kr36Platform());
        this.register(new generic_1.GenericPlatform());
    }
    register(platform) {
        this.platforms.set(platform.name, platform);
    }
    get(name) {
        return this.platforms.get(name) || this.platforms.get('generic');
    }
    detect(url) {
        for (const platform of this.platforms.values()) {
            if (platform.name === 'generic')
                continue;
            if (platform.detect(url)) {
                return platform;
            }
        }
        return this.platforms.get('generic');
    }
    list() {
        return Array.from(this.platforms.values());
    }
}
exports.PlatformRegistry = PlatformRegistry;
