// news-to-markdown/dist/platforms/segmentfault.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SegmentFaultPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class SegmentFaultPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'segmentfault';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.article-title, h1',
            author: 'meta[name="author"], .author-name, .author',
            time: 'meta[property="article:published_time"], time[datetime]',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('segmentfault.com');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.sidebar', '.recommend', '.comment',
            '.related', '.share-box',
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('.article-content, article, .post-content').first();
        if ($article.length) {
            this.processSegmentFaultContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    processSegmentFaultContent($, $content) {
        // Process images
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-src') || $img.attr('data-original');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
        // Remove code copy buttons
        $content.find('.copy-btn, .code-copy').remove();
    }
}
exports.SegmentFaultPlatform = SegmentFaultPlatform;
