// news-to-markdown/dist/platforms/toutiao.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TouTiaoPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class TouTiaoPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'toutiao';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.article__title, h1.article-title, .article-title, h1',
            author: 'meta[name="author"], .author-name, .article-author, .username, .name',
            time: 'meta[property="article:published_time"], .publish-time, time[datetime], .time',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname.replace(/^www\./, '');
            const patterns = ['toutiao.com', 'toutiao.io', '163.com', '163yun.com'];
            return patterns.some(p => hostname.includes(p));
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.nav', '.navbar', '.header', '.footer',
            '.sidebar', '.side-bar', '#sidebar',
            '.fixed-bar', '.tool-bar',
            '.pgc-off-screen',
            'script', 'style', 'noscript', 'iframe',
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // 移除头条页面底部的推荐文章区域
        this.removeRecommendations($);
        const $article = $('article').first();
        if ($article.length) {
            this.normalizeHeadings($, $article);
            this.normalizeImages($, $article);
            this.normalizeTables($, $article);
            this.normalizeBullets($, $article);
            this.normalizeCodeBlocks($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        for (const sel of ['.pgc-html', '#js_content', '.article-content']) {
            const $el = $(sel).first();
            if ($el.length) {
                this.normalizeHeadings($, $el);
                this.normalizeImages($, $el);
                this.normalizeTables($, $el);
                this.normalizeBullets($, $el);
                this.normalizeCodeBlocks($, $el);
                this.cleanEmptyTags($, $el);
                return $.html($el);
            }
        }
        const $body = $('body');
        this.normalizeHeadings($, $body);
        this.normalizeImages($, $body);
        this.normalizeTables($, $body);
        this.normalizeBullets($, $body);
        this.normalizeCodeBlocks($, $body);
        this.cleanEmptyTags($, $body);
        return $body.html() || html;
    }
    /**
     * 移除头条页面底部的推荐文章区域
     * 推荐文章在 article 外的 .item-container 中，包含 "去今日头条看专享资讯" 等链接
     */
    removeRecommendations($) {
        // 移除 article 外的所有 .item-container（推荐文章）
        $('.item-container').each((i, container) => {
            const $container = $(container);
            // 检查是否在 article 内
            if ($container.closest('article').length === 0) {
                $container.remove();
            }
        });
        // 移除 .arco-show-monitor（推荐文章容器）
        $('.arco-show-monitor').each((i, el) => {
            const $el = $(el);
            if ($el.closest('article').length === 0) {
                $el.remove();
            }
        });
        // 移除包含 "去今日头条看专享资讯" 的链接
        $('a').each((i, el) => {
            const text = $(el).text();
            if (text.includes('去今日头条看专享资讯') || text.includes('去今日头条')) {
                $(el).remove();
            }
        });
    }
    normalizeImages($, $root) {
        $root.find('div.pgc-img').each((i, wrapper) => {
            const $wrapper = $(wrapper);
            const $img = $wrapper.find('> img').first();
            const $caption = $wrapper.find('> p.pgc-img-caption').first();
            if ($img.length) {
                let src = $img.attr('data-src') || $img.attr('src') || '';
                if (src && !src.startsWith('data:image')) {
                    const alt = $caption.length
                        ? $caption.text().trim().replace(/"/g, '&#34;')
                        : ($img.attr('alt') || '').replace(/"/g, '&#34;');
                    const newImg = $(`<img src="${src}" alt="${alt}">`);
                    $wrapper.replaceWith(newImg);
                    return;
                }
            }
            $wrapper.remove();
        });
        $root.find('img').each((i, img) => {
            const $img = $(img);
            const dataSrc = $img.attr('data-src');
            if (dataSrc && !dataSrc.startsWith('data:image')) {
                $img.attr('src', dataSrc);
            }
            const src = $img.attr('src') || '';
            if (src.startsWith('data:image') || !src) {
                $img.remove();
            }
        });
    }
    normalizeHeadings($, $root) {
        let firstTitle = true;
        $root.find('[class*="pgc-h-"]').each((i, el) => {
            const cls = $(el).attr('class') || '';
            let level = 2;
            if (cls.includes('pgc-h-edit') || cls.includes('article-title')) {
                level = firstTitle ? 1 : 2;
                firstTitle = false;
            }
            else if (cls.includes('pgc-h-border-left')) {
                level = 2;
            }
            else if (cls.includes('pgc-h-forward-slash')) {
                level = 2;
            }
            const text = $(el).text().trim();
            if (!text)
                return;
            const $new = $(`<h${level}>${text}</h${level}>`);
            $(el).replaceWith($new);
        });
    }
    normalizeBullets($, $root) {
        $root.find('p').each((i, el) => {
            const text = $(el).text();
            if (text.includes('\u25cf')) {
                const items = text.split('\u25cf').map(t => t.trim()).filter(t => t);
                if (items.length > 0) {
                    const $ul = $('<ul>');
                    items.forEach(item => $ul.append($(`<li>${item}</li>`)));
                    $(el).replaceWith($ul);
                }
            }
        });
        $root.find('span').each((i, el) => {
            const text = $(el).text();
            if (text === '\u25cf' || text === '●') {
                const $next = $(el).next('span');
                if ($next.length) {
                    const content = $next.text().trim();
                    const $li = $(`<li>${content}</li>`);
                    const $ul = $(`<ul>`).append($li);
                    $(el).remove();
                    $next.replaceWith($ul);
                }
                else {
                    $(el).remove();
                }
            }
        });
    }
    normalizeCodeBlocks($, $root) {
        $root.find('pre').each((i, pre) => {
            let code = $(pre).find('> code').text() || $(pre).text();
            code = code.replace(/ {3,}/g, '\n');
            $(pre).find('> code').text(code);
        });
    }
    normalizeTables($, $root) {
        // Handle nested table wrappers: tableWrapper > syl-shadow-provider > syl-table-wrap > table
        $root.find('div.tableWrapper').each((i, wrapper) => {
            const $wrapper = $(wrapper);
            const $table = $wrapper.find('table').first(); // Find any descendant table
            if ($table.length) {
                $wrapper.replaceWith($table);
            }
            else {
                $wrapper.remove();
            }
        });
        // Handle syl-table-wrap > table
        $root.find('div.syl-table-wrap').each((i, inner) => {
            const $inner = $(inner);
            const $table = $inner.find('> table').first();
            if ($table.length) {
                $inner.replaceWith($table);
            }
            else {
                $inner.remove();
            }
        });
        // Handle other table wrappers
        $root.find('div.pgc-table, div.table-outer').each((i, wrapper) => {
            const $wrapper = $(wrapper);
            const $table = $wrapper.find('table').first(); // Find any descendant table
            if ($table.length) {
                $wrapper.replaceWith($table);
            }
            else {
                $wrapper.remove();
            }
        });
    }
    postprocess(markdown) {
        // Remove unnecessary escape characters that interfere with markdown rendering
        // Replace \- with - (escaped hyphens are not needed in most contexts)
        markdown = markdown.replace(/\\-/g, '-');
        // Remove other common unnecessary escapes
        markdown = markdown.replace(/\\\./g, '.');
        return markdown;
    }
}
exports.TouTiaoPlatform = TouTiaoPlatform;
