// news-to-markdown/dist/platforms/wechat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WechatPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class WechatPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'wechat';
        this.metaSelectors = {
            title: '#activity-name',
            author: '#js_name',
            time: 'meta[property="article:published_time"]',
        };
    }
    postprocess(markdown) {
        let result = markdown;
        // 1. 移除空图片
        result = result.replace(/!\[\]\(\s*\)/g, '');
        // 2. 移除 javascript:void(0) 链接（注意：HTML 转 Markdown 时括号会被转义）
        result = result.replace(/\[([^\]]*)\]\(javascript:void\\(0\\);\)/g, '$1');
        result = result.replace(/\[([^\]]*)\]\(javascript:void\(0\);\)/g, '$1');
        // 移除其他 javascript: 链接
        result = result.replace(/\[([^\]]*)\]\(javascript:[^)]*\)/g, '$1');
        // 3. 清理图片 URL 中的 watermark 参数和 #imgIndex
        result = result.replace(/\?from=appmsg&?watermark=\d+/g, '');
        result = result.replace(/#imgIndex=\d+/g, '');
        // 4. 确保图片独占一行（微信文章图片和图注混在同一行时分离）
        result = result.replace(/(!\[[^\]]*\]\([^)]+\))([^\n])/g, '$1\n\n$2');
        result = result.replace(/([^\n])(!\[[^\]]*\]\([^)]+\))/g, '$1\n\n$2');
        // 5. 移除 **END** 标记
        result = result.replace(/\*\*END\*\*/g, '');
        // 5. 移除文末审核信息（初审、编审、审核等）
        result = result.replace(/初审：[^\n]+\n?/g, '');
        result = result.replace(/编审：[^\n]+\n?/g, '');
        result = result.replace(/审核：[^\n]+\n?/g, '');
        result = result.replace(/供稿：[^\n]+\n?/g, '');
        // 6. 移除底部噪音
        const bottomNoisePatterns = [
            /预览时标签不可点[\s\S]*?$/gm,
            /关闭更多[\s\S]*?$/gm,
            /关闭\*\*[\s\S]*?$/gm,
            /微信扫一扫赞赏作者[\s\S]*?$/gm,
            /喜欢作者[\s\S]*?$/gm,
            /赞赏二维码[\s\S]*?$/gm,
            /搜索「」网络结[\s\S]*?$/gm,
            /返回[\s\S]*?$/gm,
            /暂无留言[\s\S]*?$/gm,
            /1条留言已无更多数据[\s\S]*?$/gm,
            /发消息[\s\S]*?$/gm,
            /写留言[\s\S]*?$/gm,
            /轻点两下[\s\S]*?$/gm,
            /调整当前正文文字大小[\s\S]*?$/gm,
            /更多\d+%$[\s\S]*?$/gm,
            /继续滑动看下一个[\s\S]*?$/gm,
            /轻触阅读原文[\s\S]*?$/gm,
            /向上滑动看下一个[\s\S]*?$/gm,
            /当前内容可能存在未经审核[\s\S]*?$/gm,
            /继续访问[\s\S]*?$/gm,
            /取消[\s\S]*?$/gm,
            /允许[\s\S]*?$/gm,
            /微信公众平台广告规范指引[\s\S]*?$/gm,
            /使用小程序[\s\S]*?$/gm,
            /使用完整服务[\s\S]*?$/gm,
            /已关注[\s\S]*?$/gm,
            /赞分享推荐[\s\S]*?$/gm,
            /分析[\s\S]*?$/gm,
            /跳转二维码[\s\S]*?$/gm,
            /作者头像[\s\S]*?$/gm,
            /微信扫一扫可打开此内容[\s\S]*?$/gm,
            /安徽省肥东县人民医院[\s\S]*?$/gm,
            /^知道了$/gm,
            /^微信扫一扫$/gm,
            /^视频$/gm,
            /^小程序$/gm,
            /^赞$/gm,
            /^在看$/gm,
            /^分享$/gm,
            /^留言$/gm,
            /^收藏$/gm,
            /^听过$/gm,
            /可在「公众号[\s\S]*?$/gm,
            /划线引导图[\s\S]*?$/gm,
            /我知道了[\s\S]*?$/gm,
            /确认提交投诉[\s\S]*?$/gm,
            /你可以补充投诉原因[\s\S]*?$/gm,
            /^确定$/gm,
        ];
        for (const pattern of bottomNoisePatterns) {
            result = result.replace(pattern, '');
        }
        // 7. 移除顶部的元信息块（作者、日期、安徽等）- 转义版本
        result = result.replace(/\[([^\]]+)\]\(javascript:[^)]+\)\s*\*(\d{4}年\d{1,2}月\d{1,2}日[^\*]*)\*\s*\*([^\*]+)\*\s*/g, '');
        result = result.replace(/\[([^\]]+)\]\(javascript:[^)]+\)\s*\*(\d{4}年\d{1,2}月\d{1,2}日[^\*]*)\*[\s\n]*/g, '');
        // 8. 移除孤立的 *安徽* 标签
        result = result.replace(/^\*安徽\*\s*$/gm, '');
        // 9. 移除顶部噪音文本
        result = result.replace(/在小说阅读器中沉浸阅读/g, '');
        // 10. 移除顶部的元信息：作者名 + 日期（转义版本）
        result = result.replace(/^[\s\n]*\[[^\]]+\]\([^)]+\)[\s\n]*\*\d{4}年\d{1,2}月\d{1,2}日[^\*]*\*[\s\n]*/gm, '');
        // 11. 移除顶部的空白标题和元信息块（转义版本）
        result = result.replace(/^#\s*\n.*?\[([^\]]+)\]\([^)]+\).*?\*(\d{4}年[^)]+)\*[\s\n]*/gm, '');
        // 12. 清理多余空行
        result = result.replace(/\n{3,}/g, '\n\n');
        // 13. 清理 HTML 实体
        result = result.replace(/&nbsp;/g, ' ');
        result = result.replace(/&amp;/g, '&');
        result = result.replace(/&lt;/g, '<');
        result = result.replace(/&gt;/g, '>');
        result = result.replace(/&quot;/g, '"');
        // 14. 移除多余的 span 标签等 HTML 残留
        result = result.replace(/<span[^>]*>/g, '');
        result = result.replace(/<\/span>/g, '');
        result = result.replace(/<br\s*\/?>/g, '\n');
        // 15. 修复图片 URL 中残留的无效参数
        result = result.replace(/mmbiz\.qpic\.cn\/[^)]*\?\)/g, 'mmbiz.qpic.cn/)');
        result = result.replace(/mmbiz\.qpic\.cn\/[^)]*&\)/g, 'mmbiz.qpic.cn/)');
        // 16. 移除顶部的 CSS 样式残留
        result = result.replace(/^[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*[\s\n]*/gm, '');
        // 17. 移除顶部的空白标题行
        result = result.replace(/^#\s*$/gm, '');
        // 18. 移除顶部残留的元信息（更 aggressive）
        // 匹配 "[作者名]" 后跟日期和地点的模式
        result = result.replace(/^\s*\[[^\]]+\]\(javascript:[^)]+\)\s*$/gm, '');
        result = result.replace(/^\s*\*\d{4}年\d{1,2}月\d{1,2}日[^\*]*\*\s*$/gm, '');
        result = result.replace(/^\s*\*[^\*]+\*\s*$/gm, '');
        // 19. 移除整个顶部元信息区域（从 # 标题后到第一段内容之前）
        const lines = result.split('\n');
        let resultLines = [];
        let foundContent = false;
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i].trim();
            // 如果遇到实际内容（非空、非图片、非装饰元素），开始保留
            if (line && !line.startsWith('![') && line.length > 10 && !line.includes('javascript:') && !line.match(/^\*[\s\S]*\*$/)) {
                foundContent = true;
            }
            // 在找到内容之前，跳过噪音行
            if (!foundContent && (line.includes('javascript:') || line.match(/^\*[\s\S]*\*$/) || !line || line === '#' || line.startsWith('['))) {
                continue;
            }
            resultLines.push(lines[i]);
        }
        result = resultLines.join('\n');
        // 20. 移除底部噪音行（纯符号、标点等）
        result = result.replace(/^[\s\n]*[，。、；：""''【】《》——……￥……]+[\s\n]*$/gm, '');
        // 21. 移除底部残留的奇怪字符
        result = result.replace(/^\[$/gm, '');
        result = result.replace(/^!\[$/gm, '');
        result = result.replace(/^##$/gm, '');
        result = result.replace(/^\[/gm, '');
        // 22. 移除底部二维码等残留图片（http:// 开头的微信图片通常是二维码/头像等装饰图）
        // 正文图片一般是 https:// 开头
        result = result.replace(/!\[.*?\]\(http:\/\/mmbiz\.qpic\.cn\/[^)]+\)/g, '');
        // 23. 移除 × 符号
        result = result.replace(/^×$/gm, '');
        // 24. 移除开头的多余换行
        result = result.replace(/^\n+/, '');
        // 25. 移除底部残留的行
        result = result.replace(/\n+!\[.*?\]\(http:\/\/mmbiz[^)]*\)\s*$/g, '');
        result = result.replace(/\n+×\s*$/g, '');
        result = result.replace(/\n+##\s*$/g, '');
        // 26. 移除文末二维码样式图片（仅处理最后一张，降低误删风险）
        // 常见为 mmbiz_png/sz_mmbiz_png，通常出现在文章末尾作为关注二维码
        result = result.replace(/\n+!\[[^\]]*\]\(https?:\/\/mmbiz\.qpic\.cn\/(?:sz_)?mmbiz_png\/[^)]+\)\s*$/g, '');
        // 27. 修复分节标题被拆分的问题（如 "**项目介绍" + 下一行独立 "**"）
        result = this.normalizeSplitSectionHeadings(result);
        // 28. 修复代码块内命令被粘连到同一行的问题
        result = this.normalizeCodeBlockCommandLines(result);
        // 29. 归一化段落换行，避免微信渲染时将多行正文合并为同一段
        result = this.normalizeParagraphBreaksForWechat(result);
        // 30. 确保标题前后有空行，提高可读性
        // 在 ## 标题前添加空行（如果前面不是空行）
        result = result.replace(/([^\n])\n(#{1,6}\s)/g, '$1\n\n$2');
        // 在 ## 标题后添加空行（如果后面不是空行）
        result = result.replace(/(#{1,6}\s[^\n]+)\n([^\n#])/g, '$1\n\n$2');
        return result.trim();
    }
    normalizeParagraphBreaksForWechat(markdown) {
        const lines = markdown.split('\n');
        const normalized = [];
        let inCodeBlock = false;
        for (let i = 0; i < lines.length; i++) {
            const current = lines[i];
            const trimmedCurrent = current.trim();
            normalized.push(current);
            if (trimmedCurrent.startsWith('```')) {
                inCodeBlock = !inCodeBlock;
                continue;
            }
            if (inCodeBlock || i === lines.length - 1)
                continue;
            const next = lines[i + 1];
            const trimmedNext = next.trim();
            if (!trimmedCurrent || !trimmedNext)
                continue;
            // 在两个纯文本行之间添加空行
            if (this.isPlainTextLine(trimmedCurrent) && this.isPlainTextLine(trimmedNext)) {
                normalized.push('');
                continue;
            }
            // 在 blockquote 后面的纯文本行之间添加空行
            if (trimmedCurrent.startsWith('>') && this.isPlainTextLine(trimmedNext)) {
                normalized.push('');
                continue;
            }
            // 在纯文本行和 blockquote 之间添加空行
            if (this.isPlainTextLine(trimmedCurrent) && trimmedNext.startsWith('>')) {
                normalized.push('');
                continue;
            }
        }
        return normalized.join('\n');
    }
    ensureBoldNumberedHeadingStandalone(markdown) {
        const lines = markdown.split('\n');
        const normalized = [];
        let inCodeBlock = false;
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            const trimmed = line.trim();
            if (trimmed.startsWith('```')) {
                inCodeBlock = !inCodeBlock;
                normalized.push(line);
                continue;
            }
            const headingMatch = trimmed.match(/^(\*\*(?:\\\(\d+\\\)|[（(]\d+[）)]|\d+\s*(?:\\?[.)]))[^\n*]*\*\*)(.*)$/);
            const isBoldNumberedHeading = !!headingMatch;
            if (inCodeBlock || !isBoldNumberedHeading) {
                normalized.push(line);
                continue;
            }
            const heading = (headingMatch?.[1] || '').trim();
            const trailingText = (headingMatch?.[2] || '').trim();
            if (normalized.length > 0 && normalized[normalized.length - 1].trim() !== '') {
                normalized.push('');
            }
            normalized.push(heading);
            if (trailingText) {
                normalized.push('');
                normalized.push(trailingText);
                continue;
            }
            const nextLine = lines[i + 1];
            if (nextLine !== undefined && nextLine.trim() !== '') {
                normalized.push('');
            }
        }
        return normalized.join('\n');
    }
    ensurePlainNumberedHeadingStandalone(markdown) {
        const lines = markdown.split('\n');
        const normalized = [];
        let inCodeBlock = false;
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            const trimmed = line.trim();
            if (trimmed.startsWith('```')) {
                inCodeBlock = !inCodeBlock;
                normalized.push(line);
                continue;
            }
            const isPlainNumberedHeading = /^(?:\\\(\d+\\\)|[（(]\d+[）)]|\d+\s*(?:\\?[.)]))\s*[^\n*]+$/.test(trimmed) && !trimmed.includes('**');
            if (inCodeBlock || !isPlainNumberedHeading) {
                normalized.push(line);
                continue;
            }
            if (normalized.length > 0 && normalized[normalized.length - 1].trim() !== '') {
                normalized.push('');
            }
            normalized.push(trimmed);
            const nextLine = lines[i + 1];
            if (nextLine !== undefined && nextLine.trim() !== '') {
                normalized.push('');
            }
        }
        return normalized.join('\n');
    }
    normalizeSplitSectionHeadings(markdown) {
        let result = markdown;
        const decoLine = '[✦◆◇★☆●○■□△▲※•·]+';
        // 把被换行拆开的粗体标题合并：
        // **项目介绍  
        // **
        // -> **项目介绍**
        result = result.replace(/\*\*([^\n*]+?)\s*\n\*\*/g, '**$1**');
        // 压缩分节编号和后续标题之间的多余空行
        result = result.replace(/(\*\*\*[^\n*]+\*\*\*)\n{2,}(\*\*[^\n]+\*\*)/g, '$1\n$2');
        // 分节编号与标题合并成同一行：
        // ***01***
        // **项目介绍**
        // -> ***01*** **项目介绍**
        result = result.replace(/^(\*\*\*[^\n*]+\*\*\*)\n(\*\*[^\n]+\*\*)$/gm, '$1 $2');
        // 分节标题与下一行粗体副标题合并：
        // ***03*** **Harness 崛起：**
        // **AI 从“能聊”走向“能干”**
        // -> ***03*** **Harness 崛起：** **AI 从“能聊”走向“能干”**
        result = result.replace(/^(\*\*\*[^\n*]+\*\*\*\s+\*\*[^\n]+\*\*)\n(\*\*[^\n]+\*\*)$/gm, '$1 $2');
        // 压缩装饰符标题块内部的空行：
        // ✦
        //
        // ✦
        // **入群申请**
        // ✦
        // -> ✦\n✦\n**入群申请**\n✦
        const decoToDeco = new RegExp(`(${decoLine})\\n{2,}(${decoLine})`, 'g');
        const decoToBold = new RegExp(`(${decoLine})\\n{2,}(\\*\\*[^\\n]+\\*\\*)`, 'g');
        const boldToDeco = new RegExp(`(\\*\\*[^\\n]+\\*\\*)\\n{2,}(${decoLine})`, 'g');
        result = result.replace(decoToDeco, '$1\n$2');
        result = result.replace(decoToBold, '$1\n$2');
        result = result.replace(boldToDeco, '$1\n$2');
        // 将装饰符标题块合并为单行（样式原文通常为同一行视觉布局）：
        // ✦\n✦\n**入群申请**\n✦ -> ✦ ✦ **入群申请** ✦
        const decoTitleInline = new RegExp(`^(${decoLine})\\n(${decoLine})\\n(\\*\\*[^\\n]+\\*\\*)\\n(${decoLine})$`, 'gm');
        result = result.replace(decoTitleInline, '$1 $2 $3 $4');
        // 确保分节标题独占一段（前后都有空行），避免与上下文被渲染为同一行
        // 适配：***01*** **项目介绍** / ***03*** **Harness 崛起：** **AI 从“能聊”走向“能干”**
        result = this.ensureSectionHeadingStandalone(result);
        // 确保普通编号标题（无粗体）独占一段
        // 适配：（5）长期记忆 / \(2\) 生成本地配置文件...
        result = this.ensurePlainNumberedHeadingStandalone(result);
        // 确保粗体编号标题独占一段（仅补缺失空行，不重复加）
        // 适配：**（3）配置要使用的模型** / **（4）为已配置的模型设置 API key**
        result = this.ensureBoldNumberedHeadingStandalone(result);
        // 修复“编号小标题 + 紧跟粗体正文”粘连
        // 例：（4）Context Engineering**子智能体...**
        result = result.replace(/((?:（|\()\d+(?:）|\))[^\n*]{1,60})\*\*([^\n].*?\*\*)/g, '$1\n**$2');
        // 修复相邻粗体片段在 Markdown 中被错误拆分的情况：
        // **...超长任务不爆上下文，**稳定跑完复杂流程。**
        // -> **...超长任务不爆上下文，稳定跑完复杂流程。**
        result = result.replace(/\*\*([^\n*]*[，。；：,:])\*\*([^\n*]+)\*\*/g, '**$1$2**');
        // 修复相邻粗体片段被写成 `****` 边界的情况：
        // **...超长任务不爆上下文，****稳定跑完复杂流程。**
        // -> **...超长任务不爆上下文，稳定跑完复杂流程。**
        result = result.replace(/\*\*([^\n*]+)\*{4}([^\n*]+)\*\*/g, '**$1$2**');
        // 修复作者/编辑徽标文本被拼成一行的问题：
        // 智猩猩AI整理编辑：没方 -> 智猩猩AI整理\n编辑：没方
        result = result.replace(/([^\n]{1,40}整理)\s*(编辑[:：])/g, '$1\n$2');
        // 确保代码块前后有正确的空行格式
        result = this.normalizeCodeBlockSpacing(result);
        return result;
    }
    ensureSectionHeadingStandalone(markdown) {
        const lines = markdown.split('\n');
        const normalized = [];
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            const trimmed = line.trim();
            const isSectionHeading = /^\*\*\*[^\n*]+\*\*\*(?:\s+\*\*[^\n]+\*\*)+$/.test(trimmed);
            if (!isSectionHeading) {
                normalized.push(line);
                continue;
            }
            if (normalized.length > 0 && normalized[normalized.length - 1].trim() !== '') {
                normalized.push('');
            }
            normalized.push(trimmed);
            const nextLine = lines[i + 1];
            if (nextLine !== undefined && nextLine.trim() !== '') {
                normalized.push('');
            }
        }
        return normalized.join('\n');
    }
    normalizeCodeBlockCommandLines(markdown) {
        const lines = markdown.split('\n');
        const normalized = [];
        let inCodeBlock = false;
        for (const rawLine of lines) {
            const trimmed = rawLine.trim();
            if (trimmed.startsWith('```')) {
                inCodeBlock = !inCodeBlock;
                normalized.push(rawLine);
                continue;
            }
            let line = rawLine;
            if (inCodeBlock && line) {
                line = line.replace(/\u00A0/g, ' ');
                // 修复 git clone 与 cd 粘连
                line = line.replace(/(\.git)\s*cd\s+/g, '$1\ncd ');
                // 把同一行里粘连的命令拆开：
                // ...）make docker-start ...  -> ...）\nmake docker-start ...
                // ...服务make down ...       -> ...服务\nmake down ...
                line = line.replace(/([^\s\/.])((?:make|npm|npx|yarn|pnpm|docker|git|python3?|pip)\s+[a-zA-Z0-9_:-]+)/g, '$1\n$2');
                // 修复被压成一行的 YAML 配置片段
                line = this.normalizeCollapsedYamlLikeLine(line);
                // 修复被压成一行的 .env 变量片段
                line = this.normalizeCollapsedEnvLine(line);
            }
            normalized.push(line);
        }
        return normalized.join('\n');
    }
    normalizeCollapsedYamlLikeLine(line) {
        if (!line.includes('models:') || !line.includes('- name:'))
            return line;
        let normalized = line;
        normalized = normalized.replace(/\s+-\s+name:/g, '\n  - name:');
        normalized = normalized.replace(/\s+(display_name:|use:|model:|api_key:|max_tokens:|temperature:|base_url:)/g, '\n  $1');
        normalized = normalized.replace(/\s+(#\s*)/g, '  $1');
        const rows = normalized.split('\n');
        const fixed = [];
        let inModels = false;
        let inItem = false;
        for (const row of rows) {
            const trimmed = row.trim();
            if (trimmed === 'models:') {
                inModels = true;
                inItem = false;
                fixed.push('models:');
                continue;
            }
            if (!inModels) {
                fixed.push(row);
                continue;
            }
            if (trimmed.startsWith('- name:')) {
                inItem = true;
                fixed.push(`  ${trimmed}`);
                continue;
            }
            if (inItem && /^(display_name:|use:|model:|api_key:|max_tokens:|temperature:|base_url:)/.test(trimmed)) {
                fixed.push(`    ${trimmed}`);
                continue;
            }
            fixed.push(trimmed ? `  ${trimmed}` : row);
        }
        return fixed.join('\n');
    }
    normalizeCollapsedEnvLine(line) {
        const envVarMatches = line.match(/[A-Z][A-Z0-9_]*=/g) || [];
        if (envVarMatches.length < 2)
            return line;
        let normalized = line;
        normalized = normalized.replace(/([a-z0-9\-])([A-Z][A-Z0-9_]*=)/g, '$1\n$2');
        normalized = normalized.replace(/(\S)(#)/g, '$1\n$2');
        normalized = normalized.replace(/(#.*?)([A-Z][A-Z0-9_]*=)/g, '$1\n$2');
        normalized = normalized.replace(/(#\s*[^#\n]+?)\s*(#\s*[^#\n]+)/g, '$1\n$2');
        return normalized;
    }
    isPlainTextLine(line) {
        if (!line)
            return false;
        if (line.startsWith('!['))
            return false;
        if (line.startsWith('#'))
            return false;
        if (line.startsWith('>'))
            return false;
        if (line.startsWith('|'))
            return false;
        if (line.startsWith('```'))
            return false;
        if (line.startsWith('- ') || line.startsWith('* '))
            return false;
        if (/^\d+\.\s/.test(line))
            return false;
        if (/^---+$/.test(line))
            return false;
        if (/^[✦◆◇★☆●○■□△▲※•·]+$/.test(line))
            return false;
        if (/^\*{2,3}.+\*{2,3}$/.test(line))
            return false;
        if (/^\*{2,3}$/.test(line))
            return false;
        return true;
    }
    /**
     * 确保代码块前后有正确的空行格式
     * 修复：
     * ```
     *
     * # 注释
     * ```
     * 变成：
     *
     * ```
     * # 注释
     * ```
     */
    normalizeCodeBlockSpacing(markdown) {
        const lines = markdown.split('\n');
        const normalized = [];
        let inCodeBlock = false;
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            const trimmed = line.trim();
            if (trimmed.startsWith('```')) {
                if (!inCodeBlock) {
                    // 代码块开始
                    // 确保代码块前有空行（如果前面不是空行）
                    if (normalized.length > 0 && normalized[normalized.length - 1].trim() !== '') {
                        normalized.push('');
                    }
                    normalized.push(line);
                    inCodeBlock = true;
                    // 跳过代码块开始后的空行
                    while (i + 1 < lines.length && lines[i + 1].trim() === '') {
                        i++;
                    }
                }
                else {
                    // 代码块结束
                    // 移除代码块结束前的空行
                    while (normalized.length > 0 && normalized[normalized.length - 1].trim() === '') {
                        normalized.pop();
                    }
                    normalized.push(line);
                    inCodeBlock = false;
                }
            }
            else {
                normalized.push(line);
            }
        }
        return normalized.join('\n');
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('mp.weixin.qq.com');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // 微信文章噪音选择器
        const noiseSelectors = [
            // 顶部/底部导航
            '#js_pc_qr_code',
            '.qr_code',
            '.weui-desktop-app__outline',
            '.mobile-page_placeholder',
            // 互动元素
            '.js_like',
            '.js_survey',
            '.js_围墙',
            '.like_area',
            '.reward_area',
            '.comment_area',
            '.share_area',
            // 广告和推广
            '[data-wechat-ad]',
            '.video_channel',
            '.article-meta',
            '.related_article',
            // 底部提示
            '.more-link',
            '.copyright_area',
            // 弹窗和遮罩
            '.wx-popup',
            '.wx-mask',
            // 图片水印
            '.watermark',
            // 视频相关
            '.video_poster',
            // 文末多余元素
            '#js_author_name',
            '.author_name',
            // 二维码相关
            '.follow_icon',
            '.wechat_avatar',
        ];
        // 移除噪音元素
        for (const selector of noiseSelectors) {
            $(selector).remove();
        }
        const $content = $('#js_content');
        if ($content.length) {
            this.flattenNestedImages($, $content);
            this.preserveSimpleSectionBoundaries($, $content);
            this.processWechatStyles($, $content);
            this.cleanEmptyTags($, $content);
            // 返回完整 HTML，让 Readability 可以识别
            return $.html();
        }
        return html;
    }
    /**
     * 展开嵌套在装饰性容器（如 135编辑器、秀米等）中的图片
     * 这些图片被深层嵌套在 section 中，Readability 可能会过滤掉
     * 策略：只处理纯装饰性容器（有图片但文本很少），用简单的图片标签替换
     */
    flattenNestedImages($, $content) {
        // 只查找带有 data-tools 属性的装饰性容器（135编辑器、秀米等）
        const decorativeSections = $content.find('section[data-tools]').toArray();
        // 按 DOM 深度排序，先处理最深层的
        decorativeSections.sort((a, b) => {
            const depthA = $(a).parents().length;
            const depthB = $(b).parents().length;
            return depthB - depthA; // 深度大的先处理
        });
        decorativeSections.forEach(el => {
            const $section = $(el);
            // 如果这个 section 已经被处理过（被父级替换了），跳过
            if (!$section.parent().length)
                return;
            // 提取容器内的所有图片
            const $images = $section.find('img');
            if ($images.length === 0)
                return;
            // 检查是否是纯装饰性容器（文本内容很少）
            const textContent = $section.text().trim();
            // 如果文本内容超过 50 个字符，说明可能包含正文，不要替换整个容器
            if (textContent.length > 50)
                return;
            // 收集有效图片
            const validImages = [];
            $images.each((j, img) => {
                const $img = $(img);
                const src = $img.attr('data-src') || $img.attr('src') || '';
                // 跳过 base64 图片和空图片
                if (!src || src.startsWith('data:image'))
                    return;
                // 注意：__bg_gif 是微信懒加载通用 class，不按 class 过滤
                // 跳过明显的小装饰图（通过 style 中的 width 判断，<= 60px）
                const style = $img.attr('style') || '';
                const widthMatch = style.match(/width:\s*(\d+)px/);
                if (widthMatch && parseInt(widthMatch[1]) <= 60)
                    return;
                // 跳过通过 data-w 属性判断的小图
                const dataW = $img.attr('data-w');
                if (dataW && parseInt(dataW) < 50)
                    return;
                validImages.push(src);
            });
            // 如果有有效图片，用简单的图片标签替换整个容器（保持原始位置）
            // 同时保留图片标题文字（如"正常的血管"等图注）
            if (validImages.length > 0) {
                let replacementHtml = validImages.map(src => `<p><img src="${src}" alt="图片"></p>`).join('\n');
                if (textContent) {
                    replacementHtml += `\n<p>${textContent}</p>`;
                }
                $section.replaceWith(replacementHtml);
            }
        });
    }
    preserveSimpleSectionBoundaries($, $content) {
        // 处理包含 blockquote 的 section：用其内容替换 section 标签，保持顺序
        $content.find('section').each((i, el) => {
            const $section = $(el);
            // 仅处理"顶层 section"
            if ($section.parents('section').length > 0)
                return;
            // 如果 section 包含 blockquote，用其子元素替换整个 section（保持顺序）
            if ($section.find('blockquote').length > 0) {
                const inner = $section.html() || '';
                $section.replaceWith(inner);
                return;
            }
        });
        $content.find('section').each((i, el) => {
            const $section = $(el);
            // 仅处理"顶层 section"（不处理嵌套 section，避免破坏复杂装饰布局）
            if ($section.parents('section').length > 0)
                return;
            // 跳过包含复杂结构的 section
            if ($section.find('section, img, table, pre, ul, ol, li, blockquote, video, iframe').length > 0)
                return;
            const text = $section.text().replace(/\s+/g, ' ').trim();
            if (!text)
                return;
            // 跳过纯装饰符号块
            if (/^[✦◆◇★☆●○■□△▲※•·\s]+$/.test(text))
                return;
            const inner = $section.html() || '';
            $section.replaceWith(`<p>${inner}</p>`);
        });
    }
    processWechatStyles($, $content) {
        $content.find('section').each((i, el) => {
            const $el = $(el);
            const style = $el.attr('style') || '';
            if (style.includes('text-align: center')) {
                $el.removeAttr('style');
            }
        });
        $content.find('img').each((i, el) => {
            const $img = $(el);
            // 1. 首先处理懒加载：用 data-src 替换 src
            const dataSrc = $img.attr('data-src');
            if (dataSrc && !dataSrc.startsWith('data:image')) {
                $img.attr('src', dataSrc);
            }
            // 2. 过滤无效图片（src 为空或 data:image）
            const src = $img.attr('src') || '';
            if (src.startsWith('data:image') || !src) {
                $img.remove();
                return;
            }
            // 注意：__bg_gif 是微信懒加载通用 class（GIF/JPEG 均可能带此 class），不按 class 过滤
            // 4. 过滤小尺寸装饰图（<= 60px）
            // 注意：只有明确设置了 width 样式的图片才会被过滤
            const style = $img.attr('style') || '';
            const widthMatch = style.match(/width:\s*(\d+)px/);
            if (widthMatch && parseInt(widthMatch[1]) <= 60) {
                $img.remove();
                return;
            }
        });
    }
}
exports.WechatPlatform = WechatPlatform;
