// news-to-markdown/dist/platforms/woshipm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WoshipmPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class WoshipmPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'woshipm';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.article-title, h1',
            author: 'meta[name="author"], .author-name, .author',
            time: 'meta[property="article:published_time"], time[datetime]',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('woshipm.com');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.sidebar', '.recommend', '.comment',
            '.related-posts', '.share-buttons',
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('article, .article-content, .post-content').first();
        if ($article.length) {
            this.processWoshipmContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    postprocess(markdown) {
        // Clean title suffix
        markdown = markdown.replace(/\s*[–-]\s*人人都是产品经理[,，]?\s*/g, '');
        return markdown;
    }
    processWoshipmContent($, $content) {
        // Process images
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-src') || $img.attr('data-original');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
    }
}
exports.WoshipmPlatform = WoshipmPlatform;
