// news-to-markdown/dist/platforms/xiaohongshu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XiaoHongShuPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class XiaoHongShuPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'xiaohongshu';
        this.metaSelectors = {
            title: 'meta[property="og:title"]',
            author: '.author-wrapper .name, .nickname',
            time: 'meta[property="article:published_time"]',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('xiaohongshu.com') || hostname.includes('xhslink.com');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        const $content = $('.note-content, .content');
        if ($content.length) {
            this.processXhsContent($, $content);
            return $.html($content);
        }
        return html;
    }
    processXhsContent($, $content) {
        $content.find('img').each((i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-src') || $img.attr('data-original');
            if (dataSrc && !dataSrc.startsWith('data:image')) {
                $img.attr('src', dataSrc);
            }
            const src = $img.attr('src') || '';
            if (src.startsWith('data:image') || !src) {
                $img.remove();
            }
        });
        $content.find('.tag').remove();
    }
}
exports.XiaoHongShuPlatform = XiaoHongShuPlatform;
