// news-to-markdown/dist/platforms/zhihu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZhihuPlatform = void 0;
const cheerio_1 = require("cheerio");
const base_1 = require("./base");
class ZhihuPlatform extends base_1.BasePlatform {
    constructor() {
        super(...arguments);
        this.name = 'zhihu';
        this.metaSelectors = {
            title: 'meta[property="og:title"], h1.Post-Title, h1',
            author: 'meta[name="author"], .UserLink-link, .AuthorInfo-name, .author',
            time: 'meta[property="article:published_time"], .ContentItem-time, time[datetime]',
        };
    }
    detect(url) {
        try {
            const hostname = new URL(url).hostname;
            return hostname.includes('zhihu.com');
        }
        catch {
            return false;
        }
    }
    preprocess(html, url) {
        const $ = (0, cheerio_1.load)(html);
        // Remove noise elements
        const noiseSelectors = [
            'header', 'footer', 'nav', 'aside',
            '.Recommendations', '.Comments', '.Modal',
            '.AppHeader', '.AppBanner',
            'script', 'style', 'noscript'
        ];
        for (const sel of noiseSelectors) {
            try {
                $(sel).remove();
            }
            catch { }
        }
        // Try to find article content
        const $article = $('.Post-RichTextContainer, .Post-RichText, .RichText, article').first();
        if ($article.length) {
            this.processZhihuContent($, $article);
            this.cleanEmptyTags($, $article);
            return $.html($article);
        }
        return html;
    }
    processZhihuContent($, $content) {
        // Process images
        $content.find('img').each((_i, el) => {
            const $img = $(el);
            const dataSrc = $img.attr('data-original') || $img.attr('data-actualsrc');
            if (dataSrc) {
                $img.attr('src', dataSrc);
            }
        });
        // Remove figure captions if they're redundant
        $content.find('figcaption').each((_i, el) => {
            const $caption = $(el);
            if ($caption.text().trim().length === 0) {
                $caption.remove();
            }
        });
    }
}
exports.ZhihuPlatform = ZhihuPlatform;
